{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import csv\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snps.ATCC_v_ElSalvador-MGYG000037768.txt\n",
      "10000\n",
      "20000\n",
      "snps.ATCC_v_UnitedStates-MGYG000092376.txt\n",
      "30000\n",
      "snps.ATCC_v_ElSalvador-MGYG000228997.txt\n",
      "40000\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866.txt\n",
      "50000\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_44_Balti.txt\n",
      "60000\n",
      "snps.ATCC_v_Mongolia-MGYG000280676.txt\n",
      "snps.ATCC_v_AM66-51ACA_human-feces_China_GCA_027667065.txt\n",
      "70000\n",
      "snps.ATCC_v_1001275B-D1_human-feces_NewYork_GCA_015558565.txt\n",
      "snps.ATCC_v_1001099B-F6_human-feces_NewYork_GCA_015554265.txt\n",
      "80000\n",
      "snps.ATCC_v_METABAT215_SUBJECTMAPPING_SCAFFOLDS_1500_MERGED_K77_METASPADES__AK_SK_5.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_5_THA1074JZ.txt\n",
      "90000\n",
      "snps.ATCC_v_2789STDY5834850_human-feces_UnitedKingdom_GCA_001406215.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000101498.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000147020.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_7_THA0058AZ.txt\n",
      "100000\n",
      "snps.ATCC_v_Fiji-MGYG000124014.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000283239.txt\n",
      "snps.ATCC_v_China-MGYG000170701.txt\n",
      "snps.ATCC_v_China-MGYG000175018.txt\n",
      "110000\n",
      "snps.ATCC_v_China-MGYG000170766.txt\n",
      "snps.ATCC_v_NCTC11814_type_GCA_900445615.txt\n",
      "snps.ATCC_v_Madagascar-MGYG000025145.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_12_Brokpa.txt\n",
      "120000\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000114505.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_A_10_1818.txt\n",
      "130000\n",
      "snps.ATCC_v_D52t1-B8_human-feces_NewYork_GCA_015558415.txt\n",
      "snps.ATCC_v_Fiji-MGYG000255607.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_8_Balti.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000185946.txt\n",
      "snps.ATCC_v_L2-32_human-feces_NA_GCA_000154085.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000073223.txt\n",
      "140000\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_5_RAU0022YZ.txt\n",
      "snps.ATCC_v_China-MGYG000287017.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000050898.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_H_3_THA1060YZ.txt\n",
      "150000\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_9_Boto.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000240385.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000021867.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_21_2500.txt\n",
      "snps.ATCC_v_iVS-1_human-feces_Nebraska_GCA_029855065.txt\n",
      "snps.ATCC_v_1001262B-human-feces_NewYork_GCA_015557685.txt\n",
      "160000\n",
      "snps.ATCC_v_Mongolia-MGYG000116659.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_11_Brokpa.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000210883.txt\n",
      "snps.ATCC_v_ca0067_human-feces_CambridgeMA_GCA_004167585.txt\n",
      "snps.ATCC_v_China-MGYG000111016.txt\n",
      "snps.ATCC_v_1001713B-A6_human-feces_NewYork_GCA_015558745.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000124287.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_21_THA1063YZ.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_15_THA1075YZ.txt\n",
      "snps.ATCC_v_LMG-10733_type_GCA_002108135.txt\n",
      "snps.ATCC_v_D53t1_human-feces_NewYork_GCA_015552825.txt\n",
      "170000\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000044941.txt\n",
      "snps.ATCC_v_Fiji-MGYG000031842.txt\n",
      "snps.ATCC_v_ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_1_Brokpa.txt\n",
      "snps.ATCC_v_Fiji-MGYG000285608.txt\n",
      "snps.ATCC_v_AF02-8_human-feces_China_GCA_027688555.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_23_THA0068JZ.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000131768.txt\n",
      "snps.ATCC_v_China-MGYG000225533.txt\n",
      "180000\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_D_23_2656.txt\n",
      "snps.ATCC_v_Madagascar-MGYG000120594.txt\n",
      "snps.ATCC_v_1001283B1-G1_human-feces_NewYork_GCA_015557595.txt\n",
      "snps.ATCC_v_BB23_human-feces_China_GCA_000817995.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000146104.txt\n",
      "snps.ATCC_v_LMG-10502_type_GCA_028743295.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_25_Purigpa.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_36_Boto.txt\n",
      "snps.ATCC_v_150_human-feces_Russia_GCA_001010915.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000056268.txt\n",
      "190000\n",
      "snps.ATCC_v_D31t1-E3_human-feces_NewYork_GCA_015549865.txt\n",
      "snps.ATCC_v_Fiji-MGYG000104252.txt\n",
      "snps.ATCC_v_AF92-02A_human-feces_China_GCA_027664845.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000132829.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000036973.txt\n",
      "snps.ATCC_v_Km4_human-feces_Russia_GCA_001756865.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_13_Brokpa.txt\n",
      "snps.ATCC_v_Madagascar-MGYG000091572.txt\n",
      "200000\n",
      "snps.ATCC_v_2789ST_UK_GCA_001406455.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000175950.txt\n",
      "snps.ATCC_v_Fiji-MGYG000099035.txt\n",
      "snps.ATCC_v_Fiji-MGYG000098158.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000099659.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000008798.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000225709.txt\n",
      "snps.ATCC_v_1001271B-A4_human-feces_NewYork_GCA_015548755.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000285386.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_21_7002.txt\n",
      "snps.ATCC_v_CLA-AP-H11_human-feces_Germany_GCA_040092935.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000138549.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000083139.txt\n",
      "snps.ATCC_v_H411_UK_GCA_001406215.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS12_B_Warli.txt\n",
      "snps.ATCC_v_D46t1-C9_human-feces_NewYork_GCA_015558575.txt\n",
      "210000\n",
      "snps.ATCC_v_China-MGYG000047373.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_37_Brokpa.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000212973.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000092433.txt\n",
      "snps.ATCC_v_AM64-16A_China_GCA_027667565.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000032473.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000066585.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS12_A_Warli.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_3_THA1067YZ_1.txt\n",
      "snps.ATCC_v_Fiji-MGYG000248641.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_P_19_2149.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_23_2501.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000124758.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000196958.txt\n",
      "snps.ATCC_v_China-MGYG000117766.txt\n",
      "snps.ATCC_v_JCM-15918_sbmtMongolia_GCA_000771705.txt\n",
      "snps.ATCC_v_China-MGYG000226008.txt\n",
      "220000\n",
      "snps.ATCC_v_AM110-65_human-feces_China_GCA_027674825.txt\n",
      "snps.ATCC_v_AM14-37_China_GCA_003472095.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000284637.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_24_1038.txt\n",
      "snps.ATCC_v_1001270B-E2_human-feces_NewYork_GCA_015548665.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000136926.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_11_RAU1009YZ.txt\n",
      "snps.ATCC_v_AF14-56_human-feces_China_GCA_003465205.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_35_Balti.txt\n",
      "snps.ATCC_v_Fiji-MGYG000076435.txt\n",
      "snps.ATCC_v_MCC257_human-feces_Ireland_GCA_018785715.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000060881.txt\n",
      "snps.ATCC_v_ZJ2-centenarian-feces_China_GCA_009832825.txt\n",
      "snps.ATCC_v_Madagascar-MGYG000012482.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000241712.txt\n",
      "snps.ATCC_v_CLA-SR-H001_human-feces_Germany_GCA_040097245.txt\n",
      "230000\n",
      "snps.ATCC_v_Mongolia-MGYG000012063.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000117819.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000059995.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_33_Brokpa.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_4_7003.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000205281.txt\n",
      "snps.ATCC_v_1-11_infant-feces_China_GCA_003030905.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_17_RAU1026YZ.txt\n",
      "snps.ATCC_v_H37_UK_GCA_001406735.txt\n",
      "snps.ATCC_v_Fiji-MGYG000199136.txt\n",
      "snps.ATCC_v_China-MGYG000093441.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS5_A_Warli.txt\n",
      "snps.ATCC_v_JCM15918_type-strain_NA_GCA_000741415.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_48_Purigpa.txt\n",
      "snps.ATCC_v_UnitedStates-MGYG000114143.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_2_Purigpa.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_45_Purigpa.txt\n",
      "240000\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000044575.txt\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_B_17_2661.txt\n",
      "snps.ATCC_v_ElSalvador-MGYG000209388.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000100616.txt\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_32_Balti.txt\n",
      "snps.ATCC_v_D59t2-181005-G7_human-feces_NewYork_GCA_015555595.txt\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000287127.txt\n",
      "snps.ATCC_v_VKPM-Ac-1245_human-feces_Russia_GCA_016069975.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000048473.txt\n",
      "snps.ATCC_v_Mongolia-MGYG000158456.txt\n"
     ]
    }
   ],
   "source": [
    "## First task is to make a list of variable positions by looping through all the SNP files.\n",
    "\n",
    "positions = []\n",
    "\n",
    "files = os.listdir(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025\")\n",
    "snpfiles =[]\n",
    "\n",
    "for i in files:\n",
    "    if i[0:4] == 'snps':\n",
    "        snpfiles.append(i)\n",
    "\n",
    "\n",
    "for j in snpfiles:\n",
    "    \n",
    "    print(j)\n",
    "    \n",
    "    with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/\"+j,\"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile, delimiter='\\t') \n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "\n",
    "\n",
    "        for row in reader:\n",
    "            row = row[0].split(' ')\n",
    "\n",
    "            # get the position and contig \n",
    "            for i in row:\n",
    "                if i != '':\n",
    "                    position = str(i)\n",
    "                    break\n",
    "                    \n",
    "            # get the ref and alt bases\n",
    "            refalt = []\n",
    "            for i in row:\n",
    "                if i in ['A','T','G','C','.']: \n",
    "                    refalt.append(i)\n",
    "                if len(refalt) == 2:\n",
    "                    break\n",
    "            \n",
    "            # ignore insertions \n",
    "            if refalt[0] != '.':\n",
    "            \n",
    "                # name and save the position \n",
    "                contig = row[-1]\n",
    "                combopos = contig+'-'+position\n",
    "\n",
    "                if combopos not in positions:\n",
    "                    positions.append(combopos)\n",
    "                    \n",
    "                    tally = len(positions)\n",
    "\n",
    "                    if tally % 10000 == 0:\n",
    "                        print(len(positions))\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "# There are 157 files. The ATCC genome is 2089645 base pairs (some of the NCBI ones are considerably larger). \n",
    "# All together there are ~250,000 putative SNP positions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "246625\n",
      "['1200', '1371', '1695', '1756', '1799', '1802', '1805', '1816', '1817', '1827']\n",
      "2076218\n",
      "\n",
      "246625\n",
      "[1200, 1371, 1695, 1756, 1799, 1802, 1805, 1816, 1817, 1827]\n",
      "2076218\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Check out current list of positions \n",
    "\n",
    "#print(len(positions))\n",
    "#print(positions[0:10])\n",
    "#print(positions[-1])\n",
    "\n",
    "\n",
    "\n",
    "### Take a look at the format of the previous positions list, to make sure same format is followed \n",
    "\n",
    "#with open('/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/position_list_101824.json', 'r') as f:\n",
    "#    positions = json.load(f)\n",
    "#print(len(positions)) # 234082\n",
    "#print(positions[0:10]) # [99, 348, 372, 549, 585, 1191, 1200, 1302, 1512, 1515]\n",
    " \n",
    "\n",
    "    \n",
    "    \n",
    "# Sp Actually, I want to remake the positions list to only be the position (no contig label required)\n",
    "positions2 = []\n",
    "for i in positions:\n",
    "    i2 = i.split('-')[1]\n",
    "    positions2.append(i2)\n",
    "print(len(positions2))\n",
    "print(positions2[0:10])\n",
    "print(positions2[-1])\n",
    "\n",
    "\n",
    "positions3 = [int(x) for x in positions2]\n",
    "print()\n",
    "print(len(positions3))\n",
    "print(positions3[0:10])\n",
    "print(positions3[-1])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# SAVE THIS LIST OF POSITIONS, since it takes a long time to make it  \n",
    "\n",
    "import json\n",
    "json_string = json.dumps(positions3)\n",
    "with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/position_list_032525.json\", \"w\") as f:\n",
    "    f.write(json_string)\n",
    "\n",
    "\n",
    "\n",
    "positions = positions3 \n",
    "\n",
    "\n",
    "# NOTE THAT THESE POSITIONS ARE NOT IN ASCENDING ORDER ACROSS THE GENOME, BECAUSE THEY WERE ASSEMBLED\n",
    "# RUNNING THROUGH NUMEROUS SAMPLES' FILES WITH ALIGNMENTS IN EITHER DIRECTION \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157\n",
      "246625\n"
     ]
    }
   ],
   "source": [
    "####  \n",
    "### \" Start the Position x Sample Table \" ###\n",
    "####\n",
    "\n",
    "# Real python people say to use lists and create the table at the end. \n",
    "# From my 161 [157] SNP files there are 234,082 [246,625] putative positions. \n",
    "# So, I'll want 161 [157] lists that are 234,082 [246,625] items long. \n",
    "# Append them to a master list called 'rows'. \n",
    "\n",
    "\n",
    "rows = []\n",
    "\n",
    "for i in range(157):\n",
    "    rows.append([None]*246625)\n",
    "\n",
    "print(len(rows))\n",
    "print(len(rows[1]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "####\n",
    "### Generate show-coords commands (b/c the .coords files show alignments)\n",
    "####\n",
    "\n",
    "path = \"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/\"\n",
    "\n",
    "deltafiles = []\n",
    "files = os.listdir(path)\n",
    "for i in files:\n",
    "    if i[-13:] == '.delta.filter':\n",
    "        deltafiles.append(i) \n",
    "        \n",
    "outfile = open(path+\"show-coords_032525.bash\",\"w\")\n",
    "\n",
    "        \n",
    "base_cmd = \"show-coords SRG_v_NewYork-1001262B.delta.filter > SRG_v_NewYork-1001262B.coords\"\n",
    "\n",
    "\n",
    "for j in deltafiles:\n",
    "    prefix = j.split('.delta.filter')[0]\n",
    "    new_cmd0 = base_cmd.replace(\"SRG_v_NewYork-1001262B.delta.filter\",j)\n",
    "    new_cmd = new_cmd0.replace(\"SRG_v_NewYork-1001262B\",prefix)\n",
    "    outfile.write(new_cmd+'\\n')\n",
    "\n",
    "outfile.close()\n",
    "\n",
    "## Then run \"bash show-coords_032525.bash\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def remove_items(test_list, item): \n",
    "    res = [i for i in test_list if i != item] \n",
    "    return res \n",
    "  \n",
    "### Define a function that will generate a list of 'missing' positions -- i.e. those not covered by an alignment. \n",
    "def inverse_ranges(ranges, min_val, max_val):\n",
    "    \"\"\"Find the inverse of a list of ranges.\n",
    "    Args:\n",
    "        ranges (list of tuples): List of ranges, where each range is a tuple (start, end).\n",
    "        min_val (int): The minimum value in the entire range.\n",
    "        max_val (int): The maximum value in the entire range.\n",
    "    Returns:\n",
    "        list of tuples: The inverse of the input ranges.\n",
    "    \"\"\"\n",
    "    inverse = []\n",
    "    last_end = min_val - 1\n",
    "    for start, end in sorted(ranges):\n",
    "        if last_end < start - 1:\n",
    "            inverse.append((last_end + 1, start - 1))\n",
    "        last_end = end\n",
    "    if last_end < max_val:\n",
    "        inverse.append((last_end + 1, max_val))\n",
    "    return inverse\n",
    "\n",
    "\n",
    "def Intersection(lst1, lst2):\n",
    "    return set(lst1).intersection(lst2)\n",
    "\n",
    "\n",
    "\n",
    "def retain_longest_non_overlapping(segments):\n",
    "    # Sort segments by their start point\n",
    "    segments.sort(key=lambda x: x[0])\n",
    "    \n",
    "    longest_segments = []\n",
    "    \n",
    "    for current in segments:\n",
    "        start, end = current\n",
    "        length = end - start\n",
    "        \n",
    "        # Check if the current segment overlaps with any in the longest_segments\n",
    "        overlaps = False\n",
    "        for longest in longest_segments:\n",
    "            l_start, l_end = longest\n",
    "            # Check for overlap\n",
    "            if not (end < l_start or start > l_end):\n",
    "                overlaps = True\n",
    "                # Retain the longest segment\n",
    "                if length > (l_end - l_start):\n",
    "                    longest_segments.remove(longest)\n",
    "                    longest_segments.append(current)\n",
    "                break\n",
    "        \n",
    "        # If no overlaps found, add the current segment\n",
    "        if not overlaps:\n",
    "            longest_segments.append(current)\n",
    "    \n",
    "    return longest_segments\n",
    "\n",
    "# Example usage\n",
    "#segments = [ (99335, 100000),(100165, 104015),(104418, 121389),(110316, 111341),  (111559, 112547), (121652, 122576), (123135, 153781)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157\n",
      "157\n",
      "['LMG-10502_type_GCA_028743295', 'Mongolia-MGYG000240385', 'AM66-51ACA_human-feces_China_GCA_027667065', '1001713B-A6_human-feces_NewYork_GCA_015558745', 'Mongolia-MGYG000100616', 'Mongolia-MGYG000012063', 'Fiji-MGYG000099035', 'Bifidobacterium-adolescentis_AK_SK_36_Boto', 'UnitedStates-MGYG000241712', 'Madagascar-MGYG000120594', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743', 'Fiji-MGYG000031842', '1001283B1-G1_human-feces_NewYork_GCA_015557595', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_A_10_1818', 'Bifidobacterium-adolescentis_AK_SK_8_Balti', 'Mongolia-MGYG000059995', 'Mongolia-MGYG000073223', 'Bifidobacterium-adolescentis_AK_SK_35_Balti', 'BB23_human-feces_China_GCA_000817995', 'Fiji-MGYG000199136', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_17_RAU1026YZ', 'Fiji-MGYG000076435', 'UnitedKingdom-MGYG000146104', '1001271B-A4_human-feces_NewYork_GCA_015548755', 'Mongolia-MGYG000048473', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_P_19_2149', 'UnitedKingdom-MGYG000287127', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_B_17_2661', 'China-MGYG000225533', 'UnitedKingdom-MGYG000083139', 'D59t2-181005-G7_human-feces_NewYork_GCA_015555595', 'UnitedStates-MGYG000147020', 'VKPM-Ac-1245_human-feces_Russia_GCA_016069975', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_7_THA0058AZ', 'JCM15918_type-strain_NA_GCA_000741415', 'UnitedKingdom-MGYG000205281', 'ElSalvador-MGYG000284637', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_5_THA1074JZ', 'Madagascar-MGYG000012482', 'AF02-8_human-feces_China_GCA_027688555', 'Bifidobacterium-adolescentis_AK_SK_45_Purigpa', 'REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_21_7002', 'ElSalvador-MGYG000210883', 'ElSalvador-MGYG000124287', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_23_2501', 'Mongolia-MGYG000008798', 'China-MGYG000117766', 'China-MGYG000287017', 'Mongolia-MGYG000021867', 'UnitedStates-MGYG000114143', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_5_RAU0022YZ', 'UnitedStates-MGYG000138549', 'ElSalvador-MGYG000228997', 'Bifidobacterium-adolescentis_AK_SK_13_Brokpa', 'D53t1_human-feces_NewYork_GCA_015552825', 'D52t1-B8_human-feces_NewYork_GCA_015558415', 'UnitedStates-MGYG000092433', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_D_23_2656', 'UnitedStates-MGYG000092376', 'Mongolia-MGYG000158456', 'REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_4_7003', 'Mongolia-MGYG000212973', 'ElSalvador-MGYG000037768', 'Bifidobacterium-adolescentis_AK_SK_2_Purigpa', 'Mongolia-MGYG000185946', 'Mongolia-MGYG000117819', '1-11_infant-feces_China_GCA_003030905', 'D31t1-E3_human-feces_NewYork_GCA_015549865', 'Mongolia-MGYG000131768', 'ElSalvador-MGYG000099659', 'UnitedKingdom-MGYG000124758', 'Bifidobacterium-adolescentis_RS5_A_Warli', 'Fiji-MGYG000248641', 'H37_UK_GCA_001406735', 'REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_24_1038', 'Fiji-MGYG000285608', 'Bifidobacterium-adolescentis_AK_SK_11_Brokpa', 'JCM-15918_sbmtMongolia_GCA_000771705', 'ZJ2-centenarian-feces_China_GCA_009832825', 'China-MGYG000111016', 'UnitedStates-MGYG000136926', 'ElSalvador-MGYG000209388', 'NCTC11814_type_GCA_900445615', 'UnitedKingdom-MGYG000044941', 'Mongolia-MGYG000060881', 'UnitedKingdom-MGYG000036973', 'iVS-1_human-feces_Nebraska_GCA_029855065', 'China-MGYG000226008', 'UnitedStates-MGYG000283239', 'Bifidobacterium-adolescentis_AK_SK_12_Brokpa', 'Bifidobacterium-adolescentis_AK_SK_33_Brokpa', '2789STDY5834850_human-feces_UnitedKingdom_GCA_001406215', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_3_THA1067YZ_1', 'ca0067_human-feces_CambridgeMA_GCA_004167585', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_H_3_THA1060YZ', 'METABAT215_SUBJECTMAPPING_SCAFFOLDS_1500_MERGED_K77_METASPADES__AK_SK_5', 'Bifidobacterium-adolescentis_RS12_A_Warli', '2789ST_UK_GCA_001406455', '1001270B-E2_human-feces_NewYork_GCA_015548665', 'ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425', 'Bifidobacterium-adolescentis_AK_SK_9_Boto', 'AF92-02A_human-feces_China_GCA_027664845', 'Mongolia-MGYG000056268', 'Bifidobacterium-adolescentis_AK_SK_48_Purigpa', 'UnitedKingdom-MGYG000044575', 'AM110-65_human-feces_China_GCA_027674825', 'UnitedKingdom-MGYG000114505', 'Bifidobacterium-adolescentis_AK_SK_37_Brokpa', 'China-MGYG000170766', 'Fiji-MGYG000255607', '150_human-feces_Russia_GCA_001010915', 'Km4_human-feces_Russia_GCA_001756865', 'Madagascar-MGYG000025145', 'UnitedKingdom-MGYG000196958', 'Bifidobacterium-adolescentis_AK_SK_44_Balti', 'Mongolia-MGYG000280676', 'AM14-37_China_GCA_003472095', 'Fiji-MGYG000104252', 'Bifidobacterium-adolescentis_AK_SK_32_Balti', 'H411_UK_GCA_001406215', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_21_THA1063YZ', 'China-MGYG000047373', 'Mongolia-MGYG000116659', 'Mongolia-MGYG000066585', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_21_2500', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_15_THA1075YZ', 'AF14-56_human-feces_China_GCA_003465205', 'L2-32_human-feces_NA_GCA_000154085', 'ElSalvador-MGYG000285386', 'China-MGYG000170701', 'MCC257_human-feces_Ireland_GCA_018785715', 'UnitedStates-MGYG000101498', 'D46t1-C9_human-feces_NewYork_GCA_015558575', '1001099B-F6_human-feces_NewYork_GCA_015554265', 'Fiji-MGYG000098158', 'China-MGYG000175018', 'Mongolia-MGYG000050898', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_11_RAU1009YZ', '1001262B-human-feces_NewYork_GCA_015557685', 'Bifidobacterium-adolescentis_AK_SK_1_Brokpa', 'LMG-10733_type_GCA_002108135', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_23_THA0068JZ', 'China-MGYG000093441', 'Fiji-MGYG000124014', 'Madagascar-MGYG000091572', 'ElSalvador-MGYG000032473', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866', '1001275B-D1_human-feces_NewYork_GCA_015558565', 'CLA-AP-H11_human-feces_Germany_GCA_040092935', 'Mongolia-MGYG000175950', 'ElSalvador-MGYG000132829', 'Bifidobacterium-adolescentis_RS12_B_Warli', 'CLA-SR-H001_human-feces_Germany_GCA_040097245', 'AM64-16A_China_GCA_027667565', 'UnitedStates-MGYG000225709', 'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ', 'Bifidobacterium-adolescentis_AK_SK_25_Purigpa']\n"
     ]
    }
   ],
   "source": [
    "coordsfiles = []\n",
    "files = os.listdir(path)\n",
    "for i in files:\n",
    "    if i[-7:] == '.coords':\n",
    "        coordsfiles.append(i) \n",
    "\n",
    "column_names = [] \n",
    "for i in coordsfiles:\n",
    "    i2 = i.split('ATCC_v_')[1]\n",
    "    i3 = i2.split('.coords')[0]\n",
    "    column_names.append(i3)\n",
    "    \n",
    "print(len(coordsfiles))\n",
    "print(len(column_names))\n",
    "print(column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "157"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coordsfiles)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATCC_v_LMG-10502_type_GCA_028743295.coords sample index:  0\n",
      "Aligned, Unaligned, Total, Percent:    2089645 0 2089645 0.0\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  0\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000240385.coords sample index:  1\n",
      "Aligned, Unaligned, Total, Percent:    1544032 545613 2089645 0.26110320174000845\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  61785\n",
      "\n",
      "\n",
      "ATCC_v_AM66-51ACA_human-feces_China_GCA_027667065.coords sample index:  2\n",
      "Aligned, Unaligned, Total, Percent:    1658576 431069 2089645 0.20628814942250956\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  37334\n",
      "\n",
      "\n",
      "ATCC_v_1001713B-A6_human-feces_NewYork_GCA_015558745.coords sample index:  3\n",
      "Aligned, Unaligned, Total, Percent:    1639743 449902 2089645 0.2153006850445889\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  39215\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000100616.coords sample index:  4\n",
      "Aligned, Unaligned, Total, Percent:    1399547 690098 2089645 0.3302465251274738\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  78369\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000012063.coords sample index:  5\n",
      "Aligned, Unaligned, Total, Percent:    1331061 758584 2089645 0.3630205130536526\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  85781\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000099035.coords sample index:  6\n",
      "Aligned, Unaligned, Total, Percent:    1576338 513307 2089645 0.24564315948402718\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  46917\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_36_Boto.coords sample index:  7\n",
      "Aligned, Unaligned, Total, Percent:    1192093 897552 2089645 0.4295236750739958\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  105157\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000241712.coords sample index:  8\n",
      "Aligned, Unaligned, Total, Percent:    1285667 803978 2089645 0.3847438201225567\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  83786\n",
      "\n",
      "\n",
      "ATCC_v_Madagascar-MGYG000120594.coords sample index:  9\n",
      "Aligned, Unaligned, Total, Percent:    1672826 416819 2089645 0.1994688092953588\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  34274\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743.coords sample index:  10\n",
      "Aligned, Unaligned, Total, Percent:    1388132 701513 2089645 0.33570917548195983\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  83785\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000031842.coords sample index:  11\n",
      "Aligned, Unaligned, Total, Percent:    1601082 488563 2089645 0.23380191372218725\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  43840\n",
      "\n",
      "\n",
      "ATCC_v_1001283B1-G1_human-feces_NewYork_GCA_015557595.coords sample index:  12\n",
      "Aligned, Unaligned, Total, Percent:    1709530 380115 2089645 0.1819041033285558\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  31409\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_A_10_1818.coords sample index:  13\n",
      "Aligned, Unaligned, Total, Percent:    1090611 999034 2089645 0.47808790488336533\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  107092\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_8_Balti.coords sample index:  14\n",
      "Aligned, Unaligned, Total, Percent:    1211497 878148 2089645 0.4202378872966461\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  97814\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000059995.coords sample index:  15\n",
      "Aligned, Unaligned, Total, Percent:    1585155 504490 2089645 0.2414237825085122\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  46868\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000073223.coords sample index:  16\n",
      "Aligned, Unaligned, Total, Percent:    1256817 832828 2089645 0.39854999294138477\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  97759\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_35_Balti.coords sample index:  17\n",
      "Aligned, Unaligned, Total, Percent:    971320 1118325 2089645 0.5351746349260281\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  135828\n",
      "\n",
      "\n",
      "ATCC_v_BB23_human-feces_China_GCA_000817995.coords sample index:  18\n",
      "Aligned, Unaligned, Total, Percent:    1715121 374524 2089645 0.1792285292477909\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  32240\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000199136.coords sample index:  19\n",
      "Aligned, Unaligned, Total, Percent:    1271735 817910 2089645 0.39141098129108054\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  94844\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_17_RAU1026YZ.coords sample index:  20\n",
      "Aligned, Unaligned, Total, Percent:    1279279 810366 2089645 0.3878007987002577\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  90444\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000076435.coords sample index:  21\n",
      "Aligned, Unaligned, Total, Percent:    1281799 807846 2089645 0.3865948522356668\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  83194\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000146104.coords sample index:  22\n",
      "Aligned, Unaligned, Total, Percent:    1436694 652951 2089645 0.3124698214289987\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  64289\n",
      "\n",
      "\n",
      "ATCC_v_1001271B-A4_human-feces_NewYork_GCA_015548755.coords sample index:  23\n",
      "Aligned, Unaligned, Total, Percent:    1650173 439472 2089645 0.2103094066216989\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  39578\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000048473.coords sample index:  24\n",
      "Aligned, Unaligned, Total, Percent:    1466073 623572 2089645 0.29841049556264343\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67857\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_P_19_2149.coords sample index:  25\n",
      "Aligned, Unaligned, Total, Percent:    848181 1241464 2089645 0.5941028260781137\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  133983\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000287127.coords sample index:  26\n",
      "Aligned, Unaligned, Total, Percent:    1638679 450966 2089645 0.2158098624407495\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  40561\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_B_17_2661.coords sample index:  27\n",
      "Aligned, Unaligned, Total, Percent:    1314599 775046 2089645 0.370898406188611\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86987\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000225533.coords sample index:  28\n",
      "Aligned, Unaligned, Total, Percent:    1505617 584028 2089645 0.2794867070722539\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  59835\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000083139.coords sample index:  29\n",
      "Aligned, Unaligned, Total, Percent:    1481586 608059 2089645 0.2909867465526441\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  62981\n",
      "\n",
      "\n",
      "ATCC_v_D59t2-181005-G7_human-feces_NewYork_GCA_015555595.coords sample index:  30\n",
      "Aligned, Unaligned, Total, Percent:    1706811 382834 2089645 0.18320528127983463\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  30750\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000147020.coords sample index:  31\n",
      "Aligned, Unaligned, Total, Percent:    1330793 758852 2089645 0.36314876450306155\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  87734\n",
      "\n",
      "\n",
      "ATCC_v_VKPM-Ac-1245_human-feces_Russia_GCA_016069975.coords sample index:  32\n",
      "Aligned, Unaligned, Total, Percent:    1680639 409006 2089645 0.1957298967049427\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  35275\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_7_THA0058AZ.coords sample index:  33\n",
      "Aligned, Unaligned, Total, Percent:    1472772 616873 2089645 0.295204687877606\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  69524\n",
      "\n",
      "\n",
      "ATCC_v_JCM15918_type-strain_NA_GCA_000741415.coords sample index:  34\n",
      "Aligned, Unaligned, Total, Percent:    1655688 433957 2089645 0.20767020235494546\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  38336\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000205281.coords sample index:  35\n",
      "Aligned, Unaligned, Total, Percent:    1441044 648601 2089645 0.3103881281270264\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  66608\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000284637.coords sample index:  36\n",
      "Aligned, Unaligned, Total, Percent:    1576558 513087 2089645 0.24553787844346767\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  46509\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_5_THA1074JZ.coords sample index:  37\n",
      "Aligned, Unaligned, Total, Percent:    1115735 973910 2089645 0.46606481005146805\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  118886\n",
      "\n",
      "\n",
      "ATCC_v_Madagascar-MGYG000012482.coords sample index:  38\n",
      "Aligned, Unaligned, Total, Percent:    1505012 584633 2089645 0.2797762299337926\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  62724\n",
      "\n",
      "\n",
      "ATCC_v_AF02-8_human-feces_China_GCA_027688555.coords sample index:  39\n",
      "Aligned, Unaligned, Total, Percent:    1673231 416414 2089645 0.19927499647069238\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  37770\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_45_Purigpa.coords sample index:  40\n",
      "Aligned, Unaligned, Total, Percent:    1337431 752214 2089645 0.35997214837927016\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  87004\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_21_7002.coords sample index:  41\n",
      "Aligned, Unaligned, Total, Percent:    1042218 1047427 2089645 0.5012463839551694\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  116309\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000210883.coords sample index:  42\n",
      "Aligned, Unaligned, Total, Percent:    1384245 705400 2089645 0.33756930004857283\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  75258\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000124287.coords sample index:  43\n",
      "Aligned, Unaligned, Total, Percent:    1452609 637036 2089645 0.3048536952448861\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  61913\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_23_2501.coords sample index:  44\n",
      "Aligned, Unaligned, Total, Percent:    1531762 557883 2089645 0.2669750125021236\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  54633\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000008798.coords sample index:  45\n",
      "Aligned, Unaligned, Total, Percent:    1536725 552920 2089645 0.26459996793713764\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  51310\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000117766.coords sample index:  46\n",
      "Aligned, Unaligned, Total, Percent:    1515124 574521 2089645 0.2749371304695295\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  54906\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000287017.coords sample index:  47\n",
      "Aligned, Unaligned, Total, Percent:    1623318 466327 2089645 0.22316087182272587\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  40505\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000021867.coords sample index:  48\n",
      "Aligned, Unaligned, Total, Percent:    1394717 694928 2089645 0.33255792251793964\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67588\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000114143.coords sample index:  49\n",
      "Aligned, Unaligned, Total, Percent:    1429558 660087 2089645 0.31588475554460205\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  71497\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_5_RAU0022YZ.coords sample index:  50\n",
      "Aligned, Unaligned, Total, Percent:    1561120 528525 2089645 0.25292573618963987\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  51095\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000138549.coords sample index:  51\n",
      "Aligned, Unaligned, Total, Percent:    1513303 576342 2089645 0.27580857035525175\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  56618\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000228997.coords sample index:  52\n",
      "Aligned, Unaligned, Total, Percent:    1348935 740710 2089645 0.35446690705837597\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  75365\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_13_Brokpa.coords sample index:  53\n",
      "Aligned, Unaligned, Total, Percent:    1245616 844029 2089645 0.4039102335564175\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  96405\n",
      "\n",
      "\n",
      "ATCC_v_D53t1_human-feces_NewYork_GCA_015552825.coords sample index:  54\n",
      "Aligned, Unaligned, Total, Percent:    1682539 407106 2089645 0.19482065135465593\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  34849\n",
      "\n",
      "\n",
      "ATCC_v_D52t1-B8_human-feces_NewYork_GCA_015558415.coords sample index:  55\n",
      "Aligned, Unaligned, Total, Percent:    1642110 447535 2089645 0.21416795675820535\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  38937\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000092433.coords sample index:  56\n",
      "Aligned, Unaligned, Total, Percent:    1634698 454947 2089645 0.21771497072469248\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  42480\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_D_23_2656.coords sample index:  57\n",
      "Aligned, Unaligned, Total, Percent:    1358345 731300 2089645 0.3499637498235346\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  73589\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000092376.coords sample index:  58\n",
      "Aligned, Unaligned, Total, Percent:    1344208 745437 2089645 0.35672901377985256\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  76708\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000158456.coords sample index:  59\n",
      "Aligned, Unaligned, Total, Percent:    1637070 452575 2089645 0.21657984968738708\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  45832\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_4_7003.coords sample index:  60\n",
      "Aligned, Unaligned, Total, Percent:    1364216 725429 2089645 0.3471541816911485\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67371\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000212973.coords sample index:  61\n",
      "Aligned, Unaligned, Total, Percent:    1289786 799859 2089645 0.3827726719131719\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86554\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000037768.coords sample index:  62\n",
      "Aligned, Unaligned, Total, Percent:    1310763 778882 2089645 0.3727341246958215\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86612\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_2_Purigpa.coords sample index:  63\n",
      "Aligned, Unaligned, Total, Percent:    1378742 710903 2089645 0.340202761713114\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  80601\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000185946.coords sample index:  64\n",
      "Aligned, Unaligned, Total, Percent:    1487886 601759 2089645 0.28797188039116695\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  61859\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000117819.coords sample index:  65\n",
      "Aligned, Unaligned, Total, Percent:    1360194 729451 2089645 0.3490789105326503\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  80090\n",
      "\n",
      "\n",
      "ATCC_v_1-11_infant-feces_China_GCA_003030905.coords sample index:  66\n",
      "Aligned, Unaligned, Total, Percent:    1679196 410449 2089645 0.1964204446209763\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  38642\n",
      "\n",
      "\n",
      "ATCC_v_D31t1-E3_human-feces_NewYork_GCA_015549865.coords sample index:  67\n",
      "Aligned, Unaligned, Total, Percent:    1687639 402006 2089645 0.1923800454144125\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  34288\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000131768.coords sample index:  68\n",
      "Aligned, Unaligned, Total, Percent:    1395464 694181 2089645 0.33220044553022166\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  72721\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000099659.coords sample index:  69\n",
      "Aligned, Unaligned, Total, Percent:    1561337 528308 2089645 0.2528218907996334\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  51271\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000124758.coords sample index:  70\n",
      "Aligned, Unaligned, Total, Percent:    1450057 639588 2089645 0.30607495531537654\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  68092\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_RS5_A_Warli.coords sample index:  71\n",
      "Aligned, Unaligned, Total, Percent:    1289045 800600 2089645 0.3831272775997837\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  83693\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000248641.coords sample index:  72\n",
      "Aligned, Unaligned, Total, Percent:    1353247 736398 2089645 0.35240339866340936\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  85685\n",
      "\n",
      "\n",
      "ATCC_v_H37_UK_GCA_001406735.coords sample index:  73\n",
      "Aligned, Unaligned, Total, Percent:    1748794 340851 2089645 0.1631143088897875\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  26502\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_24_1038.coords sample index:  74\n",
      "Aligned, Unaligned, Total, Percent:    862571 1227074 2089645 0.5872164889251523\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  133706\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000285608.coords sample index:  75\n",
      "Aligned, Unaligned, Total, Percent:    1328070 761575 2089645 0.3644518566550778\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86020\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_11_Brokpa.coords sample index:  76\n",
      "Aligned, Unaligned, Total, Percent:    1110864 978781 2089645 0.46839582799949275\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  114962\n",
      "\n",
      "\n",
      "ATCC_v_JCM-15918_sbmtMongolia_GCA_000771705.coords sample index:  77\n",
      "Aligned, Unaligned, Total, Percent:    1651697 437948 2089645 0.20958009614073203\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  39301\n",
      "\n",
      "\n",
      "ATCC_v_ZJ2-centenarian-feces_China_GCA_009832825.coords sample index:  78\n",
      "Aligned, Unaligned, Total, Percent:    1654609 435036 2089645 0.20818655800387148\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  36638\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000111016.coords sample index:  79\n",
      "Aligned, Unaligned, Total, Percent:    1456528 633117 2089645 0.3029782570723735\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67913\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000136926.coords sample index:  80\n",
      "Aligned, Unaligned, Total, Percent:    1517181 572464 2089645 0.273952752740298\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  60561\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000209388.coords sample index:  81\n",
      "Aligned, Unaligned, Total, Percent:    1397825 691820 2089645 0.33107058854494426\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  75682\n",
      "\n",
      "\n",
      "ATCC_v_NCTC11814_type_GCA_900445615.coords sample index:  82\n",
      "Aligned, Unaligned, Total, Percent:    1451872 637773 2089645 0.3052063867307605\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  81801\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000044941.coords sample index:  83\n",
      "Aligned, Unaligned, Total, Percent:    1471859 617786 2089645 0.295641604195928\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  64832\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000060881.coords sample index:  84\n",
      "Aligned, Unaligned, Total, Percent:    1244491 845154 2089645 0.40444860251382414\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  97668\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000036973.coords sample index:  85\n",
      "Aligned, Unaligned, Total, Percent:    1573427 516218 2089645 0.2470362190707034\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  52239\n",
      "\n",
      "\n",
      "ATCC_v_iVS-1_human-feces_Nebraska_GCA_029855065.coords sample index:  86\n",
      "Aligned, Unaligned, Total, Percent:    1695419 394226 2089645 0.18865692498008035\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  33990\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000226008.coords sample index:  87\n",
      "Aligned, Unaligned, Total, Percent:    1428612 661033 2089645 0.31633746401900803\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67860\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000283239.coords sample index:  88\n",
      "Aligned, Unaligned, Total, Percent:    1267130 822515 2089645 0.39361470489006506\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  92288\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_12_Brokpa.coords sample index:  89\n",
      "Aligned, Unaligned, Total, Percent:    1298060 791585 2089645 0.37881314768776514\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  84778\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_33_Brokpa.coords sample index:  90\n",
      "Aligned, Unaligned, Total, Percent:    1290617 799028 2089645 0.3823749967099675\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  93851\n",
      "\n",
      "\n",
      "ATCC_v_2789STDY5834850_human-feces_UnitedKingdom_GCA_001406215.coords sample index:  91\n",
      "Aligned, Unaligned, Total, Percent:    1679380 410265 2089645 0.19633239138705377\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  36991\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_3_THA1067YZ_1.coords sample index:  92\n",
      "Aligned, Unaligned, Total, Percent:    1410272 679373 2089645 0.3251140744001972\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  73838\n",
      "\n",
      "\n",
      "ATCC_v_ca0067_human-feces_CambridgeMA_GCA_004167585.coords sample index:  93\n",
      "Aligned, Unaligned, Total, Percent:    1695164 394481 2089645 0.18877895527709251\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  32759\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_H_3_THA1060YZ.coords sample index:  94\n",
      "Aligned, Unaligned, Total, Percent:    1500549 589096 2089645 0.2819119994065978\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  64155\n",
      "\n",
      "\n",
      "ATCC_v_METABAT215_SUBJECTMAPPING_SCAFFOLDS_1500_MERGED_K77_METASPADES__AK_SK_5.coords sample index:  95\n",
      "Aligned, Unaligned, Total, Percent:    1629934 459711 2089645 0.21999478380299045\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  42433\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_RS12_A_Warli.coords sample index:  96\n",
      "Aligned, Unaligned, Total, Percent:    1328881 760764 2089645 0.3640637524555606\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  82586\n",
      "\n",
      "\n",
      "ATCC_v_2789ST_UK_GCA_001406455.coords sample index:  97\n",
      "Aligned, Unaligned, Total, Percent:    1626463 463182 2089645 0.22165583149290907\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  43814\n",
      "\n",
      "\n",
      "ATCC_v_1001270B-E2_human-feces_NewYork_GCA_015548665.coords sample index:  98\n",
      "Aligned, Unaligned, Total, Percent:    1699047 390598 2089645 0.186920744911217\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  33297\n",
      "\n",
      "\n",
      "ATCC_v_ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425.coords sample index:  99\n",
      "Aligned, Unaligned, Total, Percent:    2089645 0 2089645 0.0\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  0\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_9_Boto.coords sample index:  100\n",
      "Aligned, Unaligned, Total, Percent:    1285430 804215 2089645 0.38485723651625037\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  84705\n",
      "\n",
      "\n",
      "ATCC_v_AF92-02A_human-feces_China_GCA_027664845.coords sample index:  101\n",
      "Aligned, Unaligned, Total, Percent:    1680238 409407 2089645 0.19592179532887166\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  34331\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000056268.coords sample index:  102\n",
      "Aligned, Unaligned, Total, Percent:    1310258 779387 2089645 0.3729757925389241\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86762\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_48_Purigpa.coords sample index:  103\n",
      "Aligned, Unaligned, Total, Percent:    1191780 897865 2089645 0.429673461281701\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  110844\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000044575.coords sample index:  104\n",
      "Aligned, Unaligned, Total, Percent:    1375370 714275 2089645 0.3418164329347808\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  77772\n",
      "\n",
      "\n",
      "ATCC_v_AM110-65_human-feces_China_GCA_027674825.coords sample index:  105\n",
      "Aligned, Unaligned, Total, Percent:    1699226 390419 2089645 0.18683508442821628\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  33153\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000114505.coords sample index:  106\n",
      "Aligned, Unaligned, Total, Percent:    1370938 718707 2089645 0.34393736735187075\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  77633\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_37_Brokpa.coords sample index:  107\n",
      "Aligned, Unaligned, Total, Percent:    1387194 702451 2089645 0.3361580555548909\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  80066\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000170766.coords sample index:  108\n",
      "Aligned, Unaligned, Total, Percent:    1411253 678392 2089645 0.3246446166693386\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  71092\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000255607.coords sample index:  109\n",
      "Aligned, Unaligned, Total, Percent:    1386489 703156 2089645 0.3364954334348657\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  71530\n",
      "\n",
      "\n",
      "ATCC_v_150_human-feces_Russia_GCA_001010915.coords sample index:  110\n",
      "Aligned, Unaligned, Total, Percent:    1678501 411144 2089645 0.1967530369991075\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  35424\n",
      "\n",
      "\n",
      "ATCC_v_Km4_human-feces_Russia_GCA_001756865.coords sample index:  111\n",
      "Aligned, Unaligned, Total, Percent:    1600754 488891 2089645 0.23395887818265781\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  41694\n",
      "\n",
      "\n",
      "ATCC_v_Madagascar-MGYG000025145.coords sample index:  112\n",
      "Aligned, Unaligned, Total, Percent:    1545118 544527 2089645 0.26058349623979193\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  54342\n",
      "\n",
      "\n",
      "ATCC_v_UnitedKingdom-MGYG000196958.coords sample index:  113\n",
      "Aligned, Unaligned, Total, Percent:    1457313 632332 2089645 0.3026025951776498\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  69382\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_44_Balti.coords sample index:  114\n",
      "Aligned, Unaligned, Total, Percent:    1410142 679503 2089645 0.32517628592416414\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  80458\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000280676.coords sample index:  115\n",
      "Aligned, Unaligned, Total, Percent:    1289999 799646 2089645 0.3826707407239029\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  96481\n",
      "\n",
      "\n",
      "ATCC_v_AM14-37_China_GCA_003472095.coords sample index:  116\n",
      "Aligned, Unaligned, Total, Percent:    1628204 461441 2089645 0.22082267562193578\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  42299\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000104252.coords sample index:  117\n",
      "Aligned, Unaligned, Total, Percent:    1543798 545847 2089645 0.26121518248314907\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  55045\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_32_Balti.coords sample index:  118\n",
      "Aligned, Unaligned, Total, Percent:    1297576 792069 2089645 0.3790447659769961\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  91681\n",
      "\n",
      "\n",
      "ATCC_v_H411_UK_GCA_001406215.coords sample index:  119\n",
      "Aligned, Unaligned, Total, Percent:    1679380 410265 2089645 0.19633239138705377\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  36991\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_21_THA1063YZ.coords sample index:  120\n",
      "Aligned, Unaligned, Total, Percent:    1510053 579592 2089645 0.27736385845442646\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  57873\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000047373.coords sample index:  121\n",
      "Aligned, Unaligned, Total, Percent:    1154412 935233 2089645 0.4475559245709199\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  107434\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000116659.coords sample index:  122\n",
      "Aligned, Unaligned, Total, Percent:    1574304 515341 2089645 0.2466165305590184\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  47575\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000066585.coords sample index:  123\n",
      "Aligned, Unaligned, Total, Percent:    1465015 624630 2089645 0.29891680165769785\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  60032\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_21_2500.coords sample index:  124\n",
      "Aligned, Unaligned, Total, Percent:    1404852 684793 2089645 0.32770781639943625\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  66609\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_15_THA1075YZ.coords sample index:  125\n",
      "Aligned, Unaligned, Total, Percent:    1027938 1061707 2089645 0.508080080587851\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  130528\n",
      "\n",
      "\n",
      "ATCC_v_AF14-56_human-feces_China_GCA_003465205.coords sample index:  126\n",
      "Aligned, Unaligned, Total, Percent:    1646661 442984 2089645 0.21199007486917634\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  41886\n",
      "\n",
      "\n",
      "ATCC_v_L2-32_human-feces_NA_GCA_000154085.coords sample index:  127\n",
      "Aligned, Unaligned, Total, Percent:    1649500 440145 2089645 0.21063147089577416\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  42506\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000285386.coords sample index:  128\n",
      "Aligned, Unaligned, Total, Percent:    1298093 791552 2089645 0.3787973555316812\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  91597\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000170701.coords sample index:  129\n",
      "Aligned, Unaligned, Total, Percent:    1459122 630523 2089645 0.30173689789413993\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  67066\n",
      "\n",
      "\n",
      "ATCC_v_MCC257_human-feces_Ireland_GCA_018785715.coords sample index:  130\n",
      "Aligned, Unaligned, Total, Percent:    1685641 404004 2089645 0.1933361886827667\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  35270\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000101498.coords sample index:  131\n",
      "Aligned, Unaligned, Total, Percent:    1619821 469824 2089645 0.22483436181743788\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  44592\n",
      "\n",
      "\n",
      "ATCC_v_D46t1-C9_human-feces_NewYork_GCA_015558575.coords sample index:  132\n",
      "Aligned, Unaligned, Total, Percent:    1639613 450032 2089645 0.2153628965685559\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  41973\n",
      "\n",
      "\n",
      "ATCC_v_1001099B-F6_human-feces_NewYork_GCA_015554265.coords sample index:  133\n",
      "Aligned, Unaligned, Total, Percent:    1578564 511081 2089645 0.24457790677363858\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  46630\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000098158.coords sample index:  134\n",
      "Aligned, Unaligned, Total, Percent:    1343005 746640 2089645 0.3573047096516394\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  84141\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000175018.coords sample index:  135\n",
      "Aligned, Unaligned, Total, Percent:    1483156 606489 2089645 0.29023542276319664\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  60921\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000050898.coords sample index:  136\n",
      "Aligned, Unaligned, Total, Percent:    1570807 518838 2089645 0.24829002055373042\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  46665\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_11_RAU1009YZ.coords sample index:  137\n",
      "Aligned, Unaligned, Total, Percent:    1617701 471944 2089645 0.2258488882082842\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  45367\n",
      "\n",
      "\n",
      "ATCC_v_1001262B-human-feces_NewYork_GCA_015557685.coords sample index:  138\n",
      "Aligned, Unaligned, Total, Percent:    1639789 449856 2089645 0.2152786717361083\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  38272\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_1_Brokpa.coords sample index:  139\n",
      "Aligned, Unaligned, Total, Percent:    1279779 809866 2089645 0.387561523608077\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  93754\n",
      "\n",
      "\n",
      "ATCC_v_LMG-10733_type_GCA_002108135.coords sample index:  140\n",
      "Aligned, Unaligned, Total, Percent:    2043897 45748 2089645 0.021892713834168003\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  4386\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_23_THA0068JZ.coords sample index:  141\n",
      "Aligned, Unaligned, Total, Percent:    1310969 778676 2089645 0.3726355433578431\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  84712\n",
      "\n",
      "\n",
      "ATCC_v_China-MGYG000093441.coords sample index:  142\n",
      "Aligned, Unaligned, Total, Percent:    1571839 517806 2089645 0.2477961567634694\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  51457\n",
      "\n",
      "\n",
      "ATCC_v_Fiji-MGYG000124014.coords sample index:  143\n",
      "Aligned, Unaligned, Total, Percent:    1574404 515241 2089645 0.24656867554058226\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  50213\n",
      "\n",
      "\n",
      "ATCC_v_Madagascar-MGYG000091572.coords sample index:  144\n",
      "Aligned, Unaligned, Total, Percent:    1312042 777603 2089645 0.3721220590100232\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  86072\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000032473.coords sample index:  145\n",
      "Aligned, Unaligned, Total, Percent:    1449147 640498 2089645 0.3065104359831455\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  66497\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866.coords sample index:  146\n",
      "Aligned, Unaligned, Total, Percent:    1689335 400310 2089645 0.19156842430173546\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  32665\n",
      "\n",
      "\n",
      "ATCC_v_1001275B-D1_human-feces_NewYork_GCA_015558565.coords sample index:  147\n",
      "Aligned, Unaligned, Total, Percent:    1677790 411855 2089645 0.1970932861801885\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  37087\n",
      "\n",
      "\n",
      "ATCC_v_CLA-AP-H11_human-feces_Germany_GCA_040092935.coords sample index:  148\n",
      "Aligned, Unaligned, Total, Percent:    1670903 418742 2089645 0.20038906129988587\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  35306\n",
      "\n",
      "\n",
      "ATCC_v_Mongolia-MGYG000175950.coords sample index:  149\n",
      "Aligned, Unaligned, Total, Percent:    1640497 449148 2089645 0.21493985820558037\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  41058\n",
      "\n",
      "\n",
      "ATCC_v_ElSalvador-MGYG000132829.coords sample index:  150\n",
      "Aligned, Unaligned, Total, Percent:    1499637 590008 2089645 0.2823484371747354\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  55721\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_RS12_B_Warli.coords sample index:  151\n",
      "Aligned, Unaligned, Total, Percent:    1237956 851689 2089645 0.40757592796862624\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  100057\n",
      "\n",
      "\n",
      "ATCC_v_CLA-SR-H001_human-feces_Germany_GCA_040097245.coords sample index:  152\n",
      "Aligned, Unaligned, Total, Percent:    1758215 331430 2089645 0.1586058876029182\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  27764\n",
      "\n",
      "\n",
      "ATCC_v_AM64-16A_China_GCA_027667565.coords sample index:  153\n",
      "Aligned, Unaligned, Total, Percent:    1660124 429521 2089645 0.20554735373711802\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  38292\n",
      "\n",
      "\n",
      "ATCC_v_UnitedStates-MGYG000225709.coords sample index:  154\n",
      "Aligned, Unaligned, Total, Percent:    1407328 682317 2089645 0.3265229261429573\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  66091\n",
      "\n",
      "\n",
      "ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ.coords sample index:  155\n",
      "Aligned, Unaligned, Total, Percent:    1109219 980426 2089645 0.4691830430527673\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  111998\n",
      "\n",
      "\n",
      "ATCC_v_Bifidobacterium-adolescentis_AK_SK_25_Purigpa.coords sample index:  156\n",
      "Aligned, Unaligned, Total, Percent:    1238062 851583 2089645 0.40752520164908396\n",
      "PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT:  96020\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "####\n",
    "###  Run through show-coords output to identify missing positions and, for each, record as N in the  position x sample allele table.\n",
    "###   \n",
    "#### \n",
    "\n",
    "\n",
    "#############################################################################################################\n",
    "#############################################################################################################\n",
    "    \n",
    "    \n",
    "    \n",
    "        ## MAKE SURE NOT TO MESS UP THIS INDEXING ##\n",
    "sample_index = -1\n",
    "for f in coordsfiles:\n",
    "        ## MAKE SURE NOT TO MESS UP THIS INDEXING ##\n",
    "        ## For every 1 into coordfiles, add 1 to sample_index\n",
    "        ## sample_index is effectively the column number of the data frame I want to make ('rows'], and it's assigned here in this loop\n",
    "        ## ok but is column names compatible with that? yes b/c it's also defined as index of coordsfiles \n",
    "          # it starts at 0 b/c we're in python\n",
    "    \n",
    "    sample_index +=1\n",
    "    rowcount = 0 \n",
    "    aligned_ranges = []\n",
    "    print(f, \"sample index: \",sample_index)\n",
    "  \n",
    "    with open(path+f,\"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile, delimiter='\\t') \n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        \n",
    "    \n",
    "        for row in reader: \n",
    "            rowcount += 1\n",
    "            row = row[0].split(' ')\n",
    "            # clean this up so it's easier to parse: ref start, ref stop, query start, q stop, r length, q length, ID, ref contig\n",
    "            row = remove_items(row,'')\n",
    "            row = remove_items(row,'|')\n",
    "\n",
    "            # save the bases that are covered by this alignment \n",
    "            refstart = min(int(row[0]),int(row[1]))\n",
    "            refend = max(int(row[0]),int(row[1]))\n",
    "            aligned_ranges.append((refstart,refend))\n",
    "            \n",
    "            # filter out repeats, keeping longest non-overlapping segments\n",
    "            aligned_ranges_filt = retain_longest_non_overlapping(aligned_ranges)\n",
    "            \n",
    "\n",
    "        \n",
    "        #### Once all the aligned ranges are defined for this sample, find the inverse -- the unaligned     \n",
    "        min_val = 1\n",
    "        max_val = 2089645 # ATCC 15703 genome length \n",
    "        unaligned_ranges = inverse_ranges(aligned_ranges_filt, min_val, max_val)\n",
    "        \n",
    "            \n",
    "        ##  calculate and print the total aligned and unaligned length\n",
    "        aligned_bp = 0\n",
    "        for r in aligned_ranges_filt:\n",
    "            l = r[1]-r[0]+1\n",
    "            aligned_bp += l\n",
    "        unaligned_bp = 0\n",
    "        for r2 in unaligned_ranges:\n",
    "            l2 = r2[1]-r2[0]+1\n",
    "            unaligned_bp += l2   \n",
    "        percent = unaligned_bp / (aligned_bp+unaligned_bp)\n",
    "        total = aligned_bp+unaligned_bp\n",
    "        print(\"Aligned, Unaligned, Total, Percent:   \",aligned_bp, unaligned_bp, total, percent)  # The total is slightly higher than expected b/c some alignments overlap\n",
    "        # gets here ^ pretty fast -- the next part is slow\n",
    "\n",
    "        \n",
    "        ## 3.  Record the unaligned positions as Ns in the list of allele tables \n",
    "        \n",
    "        # Need the sample index (range 0-157). \n",
    "        sample_list = rows[sample_index] # the list of 246625 positions \n",
    "        \n",
    "        \n",
    "        # Loop through unaligned_ranges, comparing unaligned stretches to 'positions', and recording Ns for overalps \n",
    "        for t in unaligned_ranges:\n",
    "            start = t[0]\n",
    "            end = t[1]\n",
    "            tuple_positions = list(range(start,end))            \n",
    "            # reduce size ~90% by removing all tuple_positions that aren't in 'positions' (list of 246625 putatively variable sites) \n",
    "            tp2 = Intersection(tuple_positions,positions)\n",
    "\n",
    "            ### for every unaligned tuple_position that has a SNP in some other sample\n",
    "            for p in tp2:\n",
    "                # get the index of this tuple_position in 'positions', and record the N! \n",
    "                p_index = positions.index(p)\n",
    "                sample_list[p_index] = 'N'       \n",
    "                \n",
    "\n",
    "    print(\"PUTATIVE SNP POSITIONS WITHOUT ALIGNMENT: \", sample_list.count(\"N\"))     \n",
    "    ### ONCE ALL THIS SAMPLES' N'S ARE RECORDED IN 'sample_list'\n",
    "    ### CHECK HOW MANY THERE ARE AND SAVE TO MASTER 'rows'\n",
    "    rows[sample_index] = sample_list\n",
    "    \n",
    "    # and onto the next sample\n",
    "\n",
    "    \n",
    "    print()\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 246625\n",
      "61785 246625\n",
      "37334 246625\n",
      "39215 246625\n",
      "78369 246625\n",
      "85781 246625\n",
      "46917 246625\n",
      "105157 246625\n",
      "83786 246625\n",
      "34274 246625\n",
      "83785 246625\n",
      "43840 246625\n",
      "31409 246625\n",
      "107092 246625\n",
      "97814 246625\n",
      "46868 246625\n",
      "97759 246625\n",
      "135828 246625\n",
      "32240 246625\n",
      "94844 246625\n",
      "90444 246625\n",
      "83194 246625\n",
      "64289 246625\n",
      "39578 246625\n",
      "67857 246625\n",
      "133983 246625\n",
      "40561 246625\n",
      "86987 246625\n",
      "59835 246625\n",
      "62981 246625\n",
      "30750 246625\n",
      "87734 246625\n",
      "35275 246625\n",
      "69524 246625\n",
      "38336 246625\n",
      "66608 246625\n",
      "46509 246625\n",
      "118886 246625\n",
      "62724 246625\n",
      "37770 246625\n",
      "87004 246625\n",
      "116309 246625\n",
      "75258 246625\n",
      "61913 246625\n",
      "54633 246625\n",
      "51310 246625\n",
      "54906 246625\n",
      "40505 246625\n",
      "67588 246625\n",
      "71497 246625\n",
      "51095 246625\n",
      "56618 246625\n",
      "75365 246625\n",
      "96405 246625\n",
      "34849 246625\n",
      "38937 246625\n",
      "42480 246625\n",
      "73589 246625\n",
      "76708 246625\n",
      "45832 246625\n",
      "67371 246625\n",
      "86554 246625\n",
      "86612 246625\n",
      "80601 246625\n",
      "61859 246625\n",
      "80090 246625\n",
      "38642 246625\n",
      "34288 246625\n",
      "72721 246625\n",
      "51271 246625\n",
      "68092 246625\n",
      "83693 246625\n",
      "85685 246625\n",
      "26502 246625\n",
      "133706 246625\n",
      "86020 246625\n",
      "114962 246625\n",
      "39301 246625\n",
      "36638 246625\n",
      "67913 246625\n",
      "60561 246625\n",
      "75682 246625\n",
      "81801 246625\n",
      "64832 246625\n",
      "97668 246625\n",
      "52239 246625\n",
      "33990 246625\n",
      "67860 246625\n",
      "92288 246625\n",
      "84778 246625\n",
      "93851 246625\n",
      "36991 246625\n",
      "73838 246625\n",
      "32759 246625\n",
      "64155 246625\n",
      "42433 246625\n",
      "82586 246625\n",
      "43814 246625\n",
      "33297 246625\n",
      "0 246625\n",
      "84705 246625\n",
      "34331 246625\n",
      "86762 246625\n",
      "110844 246625\n",
      "77772 246625\n",
      "33153 246625\n",
      "77633 246625\n",
      "80066 246625\n",
      "71092 246625\n",
      "71530 246625\n",
      "35424 246625\n",
      "41694 246625\n",
      "54342 246625\n",
      "69382 246625\n",
      "80458 246625\n",
      "96481 246625\n",
      "42299 246625\n",
      "55045 246625\n",
      "91681 246625\n",
      "36991 246625\n",
      "57873 246625\n",
      "107434 246625\n",
      "47575 246625\n",
      "60032 246625\n",
      "66609 246625\n",
      "130528 246625\n",
      "41886 246625\n",
      "42506 246625\n",
      "91597 246625\n",
      "67066 246625\n",
      "35270 246625\n",
      "44592 246625\n",
      "41973 246625\n",
      "46630 246625\n",
      "84141 246625\n",
      "60921 246625\n",
      "46665 246625\n",
      "45367 246625\n",
      "38272 246625\n",
      "93754 246625\n",
      "4386 246625\n",
      "84712 246625\n",
      "51457 246625\n",
      "50213 246625\n",
      "86072 246625\n",
      "66497 246625\n",
      "32665 246625\n",
      "37087 246625\n",
      "35306 246625\n",
      "41058 246625\n",
      "55721 246625\n",
      "100057 246625\n",
      "27764 246625\n",
      "38292 246625\n",
      "66091 246625\n",
      "111998 246625\n",
      "96020 246625\n"
     ]
    }
   ],
   "source": [
    "### SAVE 'rows' because it took a couple hours to put these Ns in\n",
    "for i in rows:\n",
    "    print(i.count(\"N\"), len(i))\n",
    "\n",
    "\n",
    "json_string = json.dumps(rows)\n",
    "with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/rows_vN_032525.json\", \"w\") as f:\n",
    "    f.write(json_string)\n",
    "\n",
    "\n",
    "\n",
    "# with open('/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/rows_vN_101824.json', 'r') as f:\n",
    "#     rows = json.load(f)\n",
    "\n",
    "# print(len(rows))\n",
    "# print(len(rows[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157 ['snps.ATCC_v_ElSalvador-MGYG000037768.txt', 'snps.ATCC_v_UnitedStates-MGYG000092376.txt', 'snps.ATCC_v_ElSalvador-MGYG000228997.txt', 'snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866.txt', 'snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_44_Balti.txt']\n"
     ]
    }
   ],
   "source": [
    "path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snps.ATCC_v_ElSalvador-MGYG000037768.txt\n",
      "marked N due to deletion:  1079\n",
      "marked N due to within 1000 bp of sequence end:  12890\n",
      "alt allele recorded:  26244\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000092376.txt\n",
      "marked N due to deletion:  1314\n",
      "marked N due to within 1000 bp of sequence end:  3977\n",
      "alt allele recorded:  26133\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000228997.txt\n",
      "marked N due to deletion:  1127\n",
      "marked N due to within 1000 bp of sequence end:  10682\n",
      "alt allele recorded:  26591\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866.txt\n",
      "marked N due to deletion:  1598\n",
      "marked N due to within 1000 bp of sequence end:  266\n",
      "alt allele recorded:  36220\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_44_Balti.txt\n",
      "marked N due to deletion:  641\n",
      "marked N due to within 1000 bp of sequence end:  2679\n",
      "alt allele recorded:  14978\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000280676.txt\n",
      "marked N due to deletion:  425\n",
      "marked N due to within 1000 bp of sequence end:  3248\n",
      "alt allele recorded:  10706\n",
      "\n",
      "snps.ATCC_v_AM66-51ACA_human-feces_China_GCA_027667065.txt\n",
      "marked N due to deletion:  1264\n",
      "marked N due to within 1000 bp of sequence end:  715\n",
      "alt allele recorded:  31345\n",
      "\n",
      "snps.ATCC_v_1001275B-D1_human-feces_NewYork_GCA_015558565.txt\n",
      "marked N due to deletion:  1045\n",
      "marked N due to within 1000 bp of sequence end:  683\n",
      "alt allele recorded:  29710\n",
      "\n",
      "snps.ATCC_v_1001099B-F6_human-feces_NewYork_GCA_015554265.txt\n",
      "marked N due to deletion:  1583\n",
      "marked N due to within 1000 bp of sequence end:  2041\n",
      "alt allele recorded:  31891\n",
      "\n",
      "snps.ATCC_v_METABAT215_SUBJECTMAPPING_SCAFFOLDS_1500_MERGED_K77_METASPADES__AK_SK_5.txt\n",
      "marked N due to deletion:  1252\n",
      "marked N due to within 1000 bp of sequence end:  820\n",
      "alt allele recorded:  25020\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_5_THA1074JZ.txt\n",
      "marked N due to deletion:  655\n",
      "marked N due to within 1000 bp of sequence end:  198\n",
      "alt allele recorded:  17750\n",
      "\n",
      "snps.ATCC_v_2789STDY5834850_human-feces_UnitedKingdom_GCA_001406215.txt\n",
      "marked N due to deletion:  1216\n",
      "marked N due to within 1000 bp of sequence end:  505\n",
      "alt allele recorded:  29049\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000101498.txt\n",
      "marked N due to deletion:  1620\n",
      "marked N due to within 1000 bp of sequence end:  1481\n",
      "alt allele recorded:  27211\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000147020.txt\n",
      "marked N due to deletion:  558\n",
      "marked N due to within 1000 bp of sequence end:  4430\n",
      "alt allele recorded:  13497\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_7_THA0058AZ.txt\n",
      "marked N due to deletion:  734\n",
      "marked N due to within 1000 bp of sequence end:  5293\n",
      "alt allele recorded:  19448\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000124014.txt\n",
      "marked N due to deletion:  1700\n",
      "marked N due to within 1000 bp of sequence end:  461\n",
      "alt allele recorded:  31372\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000283239.txt\n",
      "marked N due to deletion:  857\n",
      "marked N due to within 1000 bp of sequence end:  6156\n",
      "alt allele recorded:  17624\n",
      "\n",
      "snps.ATCC_v_China-MGYG000170701.txt\n",
      "marked N due to deletion:  1266\n",
      "marked N due to within 1000 bp of sequence end:  6075\n",
      "alt allele recorded:  29408\n",
      "\n",
      "snps.ATCC_v_China-MGYG000175018.txt\n",
      "marked N due to deletion:  1265\n",
      "marked N due to within 1000 bp of sequence end:  1294\n",
      "alt allele recorded:  27133\n",
      "\n",
      "snps.ATCC_v_China-MGYG000170766.txt\n",
      "marked N due to deletion:  1408\n",
      "marked N due to within 1000 bp of sequence end:  5001\n",
      "alt allele recorded:  28649\n",
      "\n",
      "snps.ATCC_v_NCTC11814_type_GCA_900445615.txt\n",
      "marked N due to deletion:  18\n",
      "marked N due to within 1000 bp of sequence end:  19\n",
      "alt allele recorded:  86\n",
      "\n",
      "snps.ATCC_v_Madagascar-MGYG000025145.txt\n",
      "marked N due to deletion:  1302\n",
      "marked N due to within 1000 bp of sequence end:  3619\n",
      "alt allele recorded:  32471\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_12_Brokpa.txt\n",
      "marked N due to deletion:  788\n",
      "marked N due to within 1000 bp of sequence end:  6676\n",
      "alt allele recorded:  23549\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000114505.txt\n",
      "marked N due to deletion:  1129\n",
      "marked N due to within 1000 bp of sequence end:  3970\n",
      "alt allele recorded:  27177\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_A_10_1818.txt\n",
      "marked N due to deletion:  1595\n",
      "marked N due to within 1000 bp of sequence end:  13365\n",
      "alt allele recorded:  30562\n",
      "\n",
      "snps.ATCC_v_D52t1-B8_human-feces_NewYork_GCA_015558415.txt\n",
      "marked N due to deletion:  1249\n",
      "marked N due to within 1000 bp of sequence end:  349\n",
      "alt allele recorded:  25970\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000255607.txt\n",
      "marked N due to deletion:  1196\n",
      "marked N due to within 1000 bp of sequence end:  13502\n",
      "alt allele recorded:  29838\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_8_Balti.txt\n",
      "marked N due to deletion:  875\n",
      "marked N due to within 1000 bp of sequence end:  7819\n",
      "alt allele recorded:  18072\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000185946.txt\n",
      "marked N due to deletion:  1208\n",
      "marked N due to within 1000 bp of sequence end:  4750\n",
      "alt allele recorded:  26089\n",
      "\n",
      "snps.ATCC_v_L2-32_human-feces_NA_GCA_000154085.txt\n",
      "marked N due to deletion:  1576\n",
      "marked N due to within 1000 bp of sequence end:  138\n",
      "alt allele recorded:  30425\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000073223.txt\n",
      "marked N due to deletion:  552\n",
      "marked N due to within 1000 bp of sequence end:  4392\n",
      "alt allele recorded:  15125\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743.txt\n",
      "marked N due to deletion:  1023\n",
      "marked N due to within 1000 bp of sequence end:  313\n",
      "alt allele recorded:  28418\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_5_RAU0022YZ.txt\n",
      "marked N due to deletion:  1063\n",
      "marked N due to within 1000 bp of sequence end:  897\n",
      "alt allele recorded:  26321\n",
      "\n",
      "snps.ATCC_v_China-MGYG000287017.txt\n",
      "marked N due to deletion:  996\n",
      "marked N due to within 1000 bp of sequence end:  456\n",
      "alt allele recorded:  19909\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000050898.txt\n",
      "marked N due to deletion:  1150\n",
      "marked N due to within 1000 bp of sequence end:  1092\n",
      "alt allele recorded:  29410\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_H_3_THA1060YZ.txt\n",
      "marked N due to deletion:  1114\n",
      "marked N due to within 1000 bp of sequence end:  2360\n",
      "alt allele recorded:  27013\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_9_Boto.txt\n",
      "marked N due to deletion:  853\n",
      "marked N due to within 1000 bp of sequence end:  5582\n",
      "alt allele recorded:  17933\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000240385.txt\n",
      "marked N due to deletion:  570\n",
      "marked N due to within 1000 bp of sequence end:  2982\n",
      "alt allele recorded:  14893\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000021867.txt\n",
      "marked N due to deletion:  1148\n",
      "marked N due to within 1000 bp of sequence end:  1087\n",
      "alt allele recorded:  21705\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_21_2500.txt\n",
      "marked N due to deletion:  1815\n",
      "marked N due to within 1000 bp of sequence end:  12087\n",
      "alt allele recorded:  36836\n",
      "\n",
      "snps.ATCC_v_iVS-1_human-feces_Nebraska_GCA_029855065.txt\n",
      "marked N due to deletion:  1274\n",
      "marked N due to within 1000 bp of sequence end:  89\n",
      "alt allele recorded:  30639\n",
      "\n",
      "snps.ATCC_v_1001262B-human-feces_NewYork_GCA_015557685.txt\n",
      "marked N due to deletion:  1400\n",
      "marked N due to within 1000 bp of sequence end:  375\n",
      "alt allele recorded:  31232\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000116659.txt\n",
      "marked N due to deletion:  1021\n",
      "marked N due to within 1000 bp of sequence end:  665\n",
      "alt allele recorded:  20098\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_11_Brokpa.txt\n",
      "marked N due to deletion:  759\n",
      "marked N due to within 1000 bp of sequence end:  9662\n",
      "alt allele recorded:  16578\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000210883.txt\n",
      "marked N due to deletion:  1024\n",
      "marked N due to within 1000 bp of sequence end:  14649\n",
      "alt allele recorded:  22669\n",
      "\n",
      "snps.ATCC_v_ca0067_human-feces_CambridgeMA_GCA_004167585.txt\n",
      "marked N due to deletion:  1243\n",
      "marked N due to within 1000 bp of sequence end:  302\n",
      "alt allele recorded:  30501\n",
      "\n",
      "snps.ATCC_v_China-MGYG000111016.txt\n",
      "marked N due to deletion:  982\n",
      "marked N due to within 1000 bp of sequence end:  6283\n",
      "alt allele recorded:  24586\n",
      "\n",
      "snps.ATCC_v_1001713B-A6_human-feces_NewYork_GCA_015558745.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "marked N due to deletion:  1672\n",
      "marked N due to within 1000 bp of sequence end:  406\n",
      "alt allele recorded:  27513\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000124287.txt\n",
      "marked N due to deletion:  1318\n",
      "marked N due to within 1000 bp of sequence end:  9798\n",
      "alt allele recorded:  30580\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_21_THA1063YZ.txt\n",
      "marked N due to deletion:  1319\n",
      "marked N due to within 1000 bp of sequence end:  2499\n",
      "alt allele recorded:  29090\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_15_THA1075YZ.txt\n",
      "marked N due to deletion:  594\n",
      "marked N due to within 1000 bp of sequence end:  177\n",
      "alt allele recorded:  16307\n",
      "\n",
      "snps.ATCC_v_LMG-10733_type_GCA_002108135.txt\n",
      "marked N due to deletion:  45\n",
      "marked N due to within 1000 bp of sequence end:  3\n",
      "alt allele recorded:  203\n",
      "\n",
      "snps.ATCC_v_D53t1_human-feces_NewYork_GCA_015552825.txt\n",
      "marked N due to deletion:  1512\n",
      "marked N due to within 1000 bp of sequence end:  433\n",
      "alt allele recorded:  32612\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000044941.txt\n",
      "marked N due to deletion:  1302\n",
      "marked N due to within 1000 bp of sequence end:  4796\n",
      "alt allele recorded:  25662\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000031842.txt\n",
      "marked N due to deletion:  1348\n",
      "marked N due to within 1000 bp of sequence end:  559\n",
      "alt allele recorded:  31542\n",
      "\n",
      "snps.ATCC_v_ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425.txt\n",
      "marked N due to deletion:  0\n",
      "marked N due to within 1000 bp of sequence end:  0\n",
      "alt allele recorded:  0\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_1_Brokpa.txt\n",
      "marked N due to deletion:  700\n",
      "marked N due to within 1000 bp of sequence end:  5893\n",
      "alt allele recorded:  18625\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000285608.txt\n",
      "marked N due to deletion:  797\n",
      "marked N due to within 1000 bp of sequence end:  6623\n",
      "alt allele recorded:  25677\n",
      "\n",
      "snps.ATCC_v_AF02-8_human-feces_China_GCA_027688555.txt\n",
      "marked N due to deletion:  1415\n",
      "marked N due to within 1000 bp of sequence end:  482\n",
      "alt allele recorded:  33934\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_23_THA0068JZ.txt\n",
      "marked N due to deletion:  1265\n",
      "marked N due to within 1000 bp of sequence end:  9206\n",
      "alt allele recorded:  28195\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000131768.txt\n",
      "marked N due to deletion:  818\n",
      "marked N due to within 1000 bp of sequence end:  5311\n",
      "alt allele recorded:  14909\n",
      "\n",
      "snps.ATCC_v_China-MGYG000225533.txt\n",
      "marked N due to deletion:  1474\n",
      "marked N due to within 1000 bp of sequence end:  1517\n",
      "alt allele recorded:  31304\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_D_23_2656.txt\n",
      "marked N due to deletion:  1767\n",
      "marked N due to within 1000 bp of sequence end:  12471\n",
      "alt allele recorded:  34267\n",
      "\n",
      "snps.ATCC_v_Madagascar-MGYG000120594.txt\n",
      "marked N due to deletion:  1670\n",
      "marked N due to within 1000 bp of sequence end:  725\n",
      "alt allele recorded:  36984\n",
      "\n",
      "snps.ATCC_v_1001283B1-G1_human-feces_NewYork_GCA_015557595.txt\n",
      "marked N due to deletion:  1257\n",
      "marked N due to within 1000 bp of sequence end:  542\n",
      "alt allele recorded:  31009\n",
      "\n",
      "snps.ATCC_v_BB23_human-feces_China_GCA_000817995.txt\n",
      "marked N due to deletion:  1342\n",
      "marked N due to within 1000 bp of sequence end:  4\n",
      "alt allele recorded:  32293\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000146104.txt\n",
      "marked N due to deletion:  1281\n",
      "marked N due to within 1000 bp of sequence end:  2825\n",
      "alt allele recorded:  28108\n",
      "\n",
      "snps.ATCC_v_LMG-10502_type_GCA_028743295.txt\n",
      "marked N due to deletion:  0\n",
      "marked N due to within 1000 bp of sequence end:  0\n",
      "alt allele recorded:  0\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_25_Purigpa.txt\n",
      "marked N due to deletion:  576\n",
      "marked N due to within 1000 bp of sequence end:  3922\n",
      "alt allele recorded:  14163\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_36_Boto.txt\n",
      "marked N due to deletion:  697\n",
      "marked N due to within 1000 bp of sequence end:  7529\n",
      "alt allele recorded:  18447\n",
      "\n",
      "snps.ATCC_v_150_human-feces_Russia_GCA_001010915.txt\n",
      "marked N due to deletion:  1422\n",
      "marked N due to within 1000 bp of sequence end:  523\n",
      "alt allele recorded:  30211\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000056268.txt\n",
      "marked N due to deletion:  807\n",
      "marked N due to within 1000 bp of sequence end:  4459\n",
      "alt allele recorded:  17712\n",
      "\n",
      "snps.ATCC_v_D31t1-E3_human-feces_NewYork_GCA_015549865.txt\n",
      "marked N due to deletion:  1495\n",
      "marked N due to within 1000 bp of sequence end:  278\n",
      "alt allele recorded:  31687\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000104252.txt\n",
      "marked N due to deletion:  1304\n",
      "marked N due to within 1000 bp of sequence end:  3087\n",
      "alt allele recorded:  31560\n",
      "\n",
      "snps.ATCC_v_AF92-02A_human-feces_China_GCA_027664845.txt\n",
      "marked N due to deletion:  1501\n",
      "marked N due to within 1000 bp of sequence end:  382\n",
      "alt allele recorded:  31390\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000132829.txt\n",
      "marked N due to deletion:  1285\n",
      "marked N due to within 1000 bp of sequence end:  2421\n",
      "alt allele recorded:  31020\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000036973.txt\n",
      "marked N due to deletion:  1373\n",
      "marked N due to within 1000 bp of sequence end:  1390\n",
      "alt allele recorded:  27220\n",
      "\n",
      "snps.ATCC_v_Km4_human-feces_Russia_GCA_001756865.txt\n",
      "marked N due to deletion:  1537\n",
      "marked N due to within 1000 bp of sequence end:  512\n",
      "alt allele recorded:  27408\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ.txt\n",
      "marked N due to deletion:  704\n",
      "marked N due to within 1000 bp of sequence end:  6743\n",
      "alt allele recorded:  16506\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_13_Brokpa.txt\n",
      "marked N due to deletion:  859\n",
      "marked N due to within 1000 bp of sequence end:  7975\n",
      "alt allele recorded:  17110\n",
      "\n",
      "snps.ATCC_v_Madagascar-MGYG000091572.txt\n",
      "marked N due to deletion:  1028\n",
      "marked N due to within 1000 bp of sequence end:  7433\n",
      "alt allele recorded:  28358\n",
      "\n",
      "snps.ATCC_v_2789ST_UK_GCA_001406455.txt\n",
      "marked N due to deletion:  1364\n",
      "marked N due to within 1000 bp of sequence end:  257\n",
      "alt allele recorded:  30892\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000175950.txt\n",
      "marked N due to deletion:  1157\n",
      "marked N due to within 1000 bp of sequence end:  1153\n",
      "alt allele recorded:  29829\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000099035.txt\n",
      "marked N due to deletion:  1264\n",
      "marked N due to within 1000 bp of sequence end:  378\n",
      "alt allele recorded:  32420\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000098158.txt\n",
      "marked N due to deletion:  837\n",
      "marked N due to within 1000 bp of sequence end:  9240\n",
      "alt allele recorded:  22787\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000099659.txt\n",
      "marked N due to deletion:  1319\n",
      "marked N due to within 1000 bp of sequence end:  3318\n",
      "alt allele recorded:  30843\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000008798.txt\n",
      "marked N due to deletion:  1163\n",
      "marked N due to within 1000 bp of sequence end:  1688\n",
      "alt allele recorded:  29461\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000225709.txt\n",
      "marked N due to deletion:  1373\n",
      "marked N due to within 1000 bp of sequence end:  5544\n",
      "alt allele recorded:  27141\n",
      "\n",
      "snps.ATCC_v_1001271B-A4_human-feces_NewYork_GCA_015548755.txt\n",
      "marked N due to deletion:  1492\n",
      "marked N due to within 1000 bp of sequence end:  465\n",
      "alt allele recorded:  30542\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000285386.txt\n",
      "marked N due to deletion:  1057\n",
      "marked N due to within 1000 bp of sequence end:  12468\n",
      "alt allele recorded:  24597\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_21_7002.txt\n",
      "marked N due to deletion:  750\n",
      "marked N due to within 1000 bp of sequence end:  426\n",
      "alt allele recorded:  17148\n",
      "\n",
      "snps.ATCC_v_CLA-AP-H11_human-feces_Germany_GCA_040092935.txt\n",
      "marked N due to deletion:  1606\n",
      "marked N due to within 1000 bp of sequence end:  518\n",
      "alt allele recorded:  31269\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000138549.txt\n",
      "marked N due to deletion:  2419\n",
      "marked N due to within 1000 bp of sequence end:  6372\n",
      "alt allele recorded:  28118\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000083139.txt\n",
      "marked N due to deletion:  1347\n",
      "marked N due to within 1000 bp of sequence end:  2595\n",
      "alt allele recorded:  28016\n",
      "\n",
      "snps.ATCC_v_H411_UK_GCA_001406215.txt\n",
      "marked N due to deletion:  1216\n",
      "marked N due to within 1000 bp of sequence end:  505\n",
      "alt allele recorded:  29049\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS12_B_Warli.txt\n",
      "marked N due to deletion:  1205\n",
      "marked N due to within 1000 bp of sequence end:  6823\n",
      "alt allele recorded:  19890\n",
      "\n",
      "snps.ATCC_v_D46t1-C9_human-feces_NewYork_GCA_015558575.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "marked N due to deletion:  1512\n",
      "marked N due to within 1000 bp of sequence end:  458\n",
      "alt allele recorded:  32585\n",
      "\n",
      "snps.ATCC_v_China-MGYG000047373.txt\n",
      "marked N due to deletion:  809\n",
      "marked N due to within 1000 bp of sequence end:  7279\n",
      "alt allele recorded:  16180\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_37_Brokpa.txt\n",
      "marked N due to deletion:  853\n",
      "marked N due to within 1000 bp of sequence end:  6113\n",
      "alt allele recorded:  18498\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000212973.txt\n",
      "marked N due to deletion:  1098\n",
      "marked N due to within 1000 bp of sequence end:  3576\n",
      "alt allele recorded:  22456\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000092433.txt\n",
      "marked N due to deletion:  1615\n",
      "marked N due to within 1000 bp of sequence end:  1872\n",
      "alt allele recorded:  30166\n",
      "\n",
      "snps.ATCC_v_AM64-16A_China_GCA_027667565.txt\n",
      "marked N due to deletion:  1304\n",
      "marked N due to within 1000 bp of sequence end:  700\n",
      "alt allele recorded:  30781\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000032473.txt\n",
      "marked N due to deletion:  1369\n",
      "marked N due to within 1000 bp of sequence end:  8207\n",
      "alt allele recorded:  30987\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000066585.txt\n",
      "marked N due to deletion:  903\n",
      "marked N due to within 1000 bp of sequence end:  2909\n",
      "alt allele recorded:  18744\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS12_A_Warli.txt\n",
      "marked N due to deletion:  1148\n",
      "marked N due to within 1000 bp of sequence end:  2825\n",
      "alt allele recorded:  23998\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_3_THA1067YZ_1.txt\n",
      "marked N due to deletion:  970\n",
      "marked N due to within 1000 bp of sequence end:  5869\n",
      "alt allele recorded:  23392\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000248641.txt\n",
      "marked N due to deletion:  824\n",
      "marked N due to within 1000 bp of sequence end:  10013\n",
      "alt allele recorded:  23897\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_P_19_2149.txt\n",
      "marked N due to deletion:  982\n",
      "marked N due to within 1000 bp of sequence end:  178\n",
      "alt allele recorded:  21342\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_23_2501.txt\n",
      "marked N due to deletion:  1854\n",
      "marked N due to within 1000 bp of sequence end:  3016\n",
      "alt allele recorded:  39474\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000124758.txt\n",
      "marked N due to deletion:  1080\n",
      "marked N due to within 1000 bp of sequence end:  3209\n",
      "alt allele recorded:  26625\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000196958.txt\n",
      "marked N due to deletion:  1127\n",
      "marked N due to within 1000 bp of sequence end:  4418\n",
      "alt allele recorded:  27942\n",
      "\n",
      "snps.ATCC_v_China-MGYG000117766.txt\n",
      "marked N due to deletion:  1245\n",
      "marked N due to within 1000 bp of sequence end:  3470\n",
      "alt allele recorded:  33192\n",
      "\n",
      "snps.ATCC_v_JCM-15918_sbmtMongolia_GCA_000771705.txt\n",
      "marked N due to deletion:  1410\n",
      "marked N due to within 1000 bp of sequence end:  699\n",
      "alt allele recorded:  31241\n",
      "\n",
      "snps.ATCC_v_China-MGYG000226008.txt\n",
      "marked N due to deletion:  1208\n",
      "marked N due to within 1000 bp of sequence end:  2918\n",
      "alt allele recorded:  23243\n",
      "\n",
      "snps.ATCC_v_AM110-65_human-feces_China_GCA_027674825.txt\n",
      "marked N due to deletion:  1279\n",
      "marked N due to within 1000 bp of sequence end:  538\n",
      "alt allele recorded:  31701\n",
      "\n",
      "snps.ATCC_v_AM14-37_China_GCA_003472095.txt\n",
      "marked N due to deletion:  1494\n",
      "marked N due to within 1000 bp of sequence end:  272\n",
      "alt allele recorded:  30344\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000284637.txt\n",
      "marked N due to deletion:  1357\n",
      "marked N due to within 1000 bp of sequence end:  949\n",
      "alt allele recorded:  31583\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_24_1038.txt\n",
      "marked N due to deletion:  637\n",
      "marked N due to within 1000 bp of sequence end:  96\n",
      "alt allele recorded:  14488\n",
      "\n",
      "snps.ATCC_v_1001270B-E2_human-feces_NewYork_GCA_015548665.txt\n",
      "marked N due to deletion:  1381\n",
      "marked N due to within 1000 bp of sequence end:  460\n",
      "alt allele recorded:  30707\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000136926.txt\n",
      "marked N due to deletion:  1248\n",
      "marked N due to within 1000 bp of sequence end:  2018\n",
      "alt allele recorded:  27103\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_11_RAU1009YZ.txt\n",
      "marked N due to deletion:  1265\n",
      "marked N due to within 1000 bp of sequence end:  764\n",
      "alt allele recorded:  28787\n",
      "\n",
      "snps.ATCC_v_AF14-56_human-feces_China_GCA_003465205.txt\n",
      "marked N due to deletion:  1331\n",
      "marked N due to within 1000 bp of sequence end:  680\n",
      "alt allele recorded:  32788\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_35_Balti.txt\n",
      "marked N due to deletion:  587\n",
      "marked N due to within 1000 bp of sequence end:  5674\n",
      "alt allele recorded:  12398\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000076435.txt\n",
      "marked N due to deletion:  1264\n",
      "marked N due to within 1000 bp of sequence end:  7295\n",
      "alt allele recorded:  31380\n",
      "\n",
      "snps.ATCC_v_MCC257_human-feces_Ireland_GCA_018785715.txt\n",
      "marked N due to deletion:  1433\n",
      "marked N due to within 1000 bp of sequence end:  502\n",
      "alt allele recorded:  30328\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000060881.txt\n",
      "marked N due to deletion:  752\n",
      "marked N due to within 1000 bp of sequence end:  6175\n",
      "alt allele recorded:  15754\n",
      "\n",
      "snps.ATCC_v_ZJ2-centenarian-feces_China_GCA_009832825.txt\n",
      "marked N due to deletion:  1297\n",
      "marked N due to within 1000 bp of sequence end:  7\n",
      "alt allele recorded:  34161\n",
      "\n",
      "snps.ATCC_v_Madagascar-MGYG000012482.txt\n",
      "marked N due to deletion:  1345\n",
      "marked N due to within 1000 bp of sequence end:  5920\n",
      "alt allele recorded:  32459\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000241712.txt\n",
      "marked N due to deletion:  882\n",
      "marked N due to within 1000 bp of sequence end:  5809\n",
      "alt allele recorded:  16370\n",
      "\n",
      "snps.ATCC_v_CLA-SR-H001_human-feces_Germany_GCA_040097245.txt\n",
      "marked N due to deletion:  1642\n",
      "marked N due to within 1000 bp of sequence end:  846\n",
      "alt allele recorded:  31974\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000012063.txt\n",
      "marked N due to deletion:  921\n",
      "marked N due to within 1000 bp of sequence end:  5382\n",
      "alt allele recorded:  17767\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000117819.txt\n",
      "marked N due to deletion:  951\n",
      "marked N due to within 1000 bp of sequence end:  6220\n",
      "alt allele recorded:  16201\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000059995.txt\n",
      "marked N due to deletion:  988\n",
      "marked N due to within 1000 bp of sequence end:  875\n",
      "alt allele recorded:  19665\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_33_Brokpa.txt\n",
      "marked N due to deletion:  540\n",
      "marked N due to within 1000 bp of sequence end:  5925\n",
      "alt allele recorded:  16126\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_4_7003.txt\n",
      "marked N due to deletion:  1361\n",
      "marked N due to within 1000 bp of sequence end:  599\n",
      "alt allele recorded:  27206\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000205281.txt\n",
      "marked N due to deletion:  1952\n",
      "marked N due to within 1000 bp of sequence end:  6818\n",
      "alt allele recorded:  29305\n",
      "\n",
      "snps.ATCC_v_1-11_infant-feces_China_GCA_003030905.txt\n",
      "marked N due to deletion:  895\n",
      "marked N due to within 1000 bp of sequence end:  3\n",
      "alt allele recorded:  20382\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_17_RAU1026YZ.txt\n",
      "marked N due to deletion:  794\n",
      "marked N due to within 1000 bp of sequence end:  7687\n",
      "alt allele recorded:  20516\n",
      "\n",
      "snps.ATCC_v_H37_UK_GCA_001406735.txt\n",
      "marked N due to deletion:  1569\n",
      "marked N due to within 1000 bp of sequence end:  665\n",
      "alt allele recorded:  31955\n",
      "\n",
      "snps.ATCC_v_Fiji-MGYG000199136.txt\n",
      "marked N due to deletion:  1011\n",
      "marked N due to within 1000 bp of sequence end:  7805\n",
      "alt allele recorded:  24889\n",
      "\n",
      "snps.ATCC_v_China-MGYG000093441.txt\n",
      "marked N due to deletion:  1573\n",
      "marked N due to within 1000 bp of sequence end:  3093\n",
      "alt allele recorded:  28763\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_RS5_A_Warli.txt\n",
      "marked N due to deletion:  1066\n",
      "marked N due to within 1000 bp of sequence end:  8201\n",
      "alt allele recorded:  26035\n",
      "\n",
      "snps.ATCC_v_JCM15918_type-strain_NA_GCA_000741415.txt\n",
      "marked N due to deletion:  1440\n",
      "marked N due to within 1000 bp of sequence end:  364\n",
      "alt allele recorded:  31375\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_48_Purigpa.txt\n",
      "marked N due to deletion:  498\n",
      "marked N due to within 1000 bp of sequence end:  5778\n",
      "alt allele recorded:  12200\n",
      "\n",
      "snps.ATCC_v_UnitedStates-MGYG000114143.txt\n",
      "marked N due to deletion:  1185\n",
      "marked N due to within 1000 bp of sequence end:  5166\n",
      "alt allele recorded:  23871\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_2_Purigpa.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "marked N due to deletion:  697\n",
      "marked N due to within 1000 bp of sequence end:  5437\n",
      "alt allele recorded:  17086\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_45_Purigpa.txt\n",
      "marked N due to deletion:  679\n",
      "marked N due to within 1000 bp of sequence end:  3668\n",
      "alt allele recorded:  16432\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000044575.txt\n",
      "marked N due to deletion:  2068\n",
      "marked N due to within 1000 bp of sequence end:  10173\n",
      "alt allele recorded:  26475\n",
      "\n",
      "snps.ATCC_v_REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_B_17_2661.txt\n",
      "marked N due to deletion:  1506\n",
      "marked N due to within 1000 bp of sequence end:  9030\n",
      "alt allele recorded:  33220\n",
      "\n",
      "snps.ATCC_v_ElSalvador-MGYG000209388.txt\n",
      "marked N due to deletion:  1102\n",
      "marked N due to within 1000 bp of sequence end:  7188\n",
      "alt allele recorded:  25607\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000100616.txt\n",
      "marked N due to deletion:  1130\n",
      "marked N due to within 1000 bp of sequence end:  5223\n",
      "alt allele recorded:  19547\n",
      "\n",
      "snps.ATCC_v_Bifidobacterium-adolescentis_AK_SK_32_Balti.txt\n",
      "marked N due to deletion:  715\n",
      "marked N due to within 1000 bp of sequence end:  2836\n",
      "alt allele recorded:  16213\n",
      "\n",
      "snps.ATCC_v_D59t2-181005-G7_human-feces_NewYork_GCA_015555595.txt\n",
      "marked N due to deletion:  1533\n",
      "marked N due to within 1000 bp of sequence end:  344\n",
      "alt allele recorded:  32389\n",
      "\n",
      "snps.ATCC_v_UnitedKingdom-MGYG000287127.txt\n",
      "marked N due to deletion:  1334\n",
      "marked N due to within 1000 bp of sequence end:  1007\n",
      "alt allele recorded:  29548\n",
      "\n",
      "snps.ATCC_v_VKPM-Ac-1245_human-feces_Russia_GCA_016069975.txt\n",
      "marked N due to deletion:  992\n",
      "marked N due to within 1000 bp of sequence end:  334\n",
      "alt allele recorded:  22164\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000048473.txt\n",
      "marked N due to deletion:  1016\n",
      "marked N due to within 1000 bp of sequence end:  6226\n",
      "alt allele recorded:  18505\n",
      "\n",
      "snps.ATCC_v_Mongolia-MGYG000158456.txt\n",
      "marked N due to deletion:  1184\n",
      "marked N due to within 1000 bp of sequence end:  1666\n",
      "alt allele recorded:  23331\n",
      "\n"
     ]
    }
   ],
   "source": [
    "####\n",
    "###  Run through SNP files again to fill in ALT alleles. If the sample has a deletion, record as N.  \n",
    "###  If the SNP is within 1000 bp of the nearest sequence end [DIST], record as N. \n",
    "### \n",
    "#### \n",
    "\n",
    "\n",
    "\n",
    "files = os.listdir(path)\n",
    "snpfiles =[]\n",
    "for i in files:\n",
    "    if i[0:4] == 'snps':\n",
    "        snpfiles.append(i)\n",
    "        \n",
    "        \n",
    "for j in snpfiles:\n",
    "    \n",
    "    print(j) \n",
    "\n",
    "    col_name = j.split('snps.ATCC_v_')[1].split('.txt')[0] ## this might be an issue ... don't remake col_name??\n",
    "                                                           # actually it's OK b/c col_name is used to find the right index of column_names later\n",
    "    alt_delete = 0\n",
    "    alt_distance = 0\n",
    "    alt_alt = 0\n",
    "\n",
    "    with open(path+j,\"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile, delimiter='\\t') \n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "        header=next(reader)\n",
    "\n",
    "        for row in reader:\n",
    "            row = row[0].split(' ') # clean this up so it's easier to parse: \n",
    "            # ref start, ref stop, query start, q stop, r length, q length, ID, ref contig\n",
    "            row = remove_items(row,'')\n",
    "            row = remove_items(row,'|')  \n",
    "\n",
    "            pos = int(row[0]) # position in ref genome \n",
    "            alt = row[2] # alt allele\n",
    "            ref = row[1]\n",
    "          #  buff = row[4] # [BUFF] distance from this SNP to the nearest mismatch (end of alignment, indel, SNP, etc) in the same alignment\n",
    "            dist = int(row[5]) # [DIST] distance from this SNP to the nearest sequence end \n",
    "\n",
    "\n",
    "            ### Should this alt allele be saved? Three requirements:\n",
    "            # 1. If the sample has a deletion, record as N.  \n",
    "            if alt == '.':\n",
    "                alt = 'N'\n",
    "                alt_delete += 1\n",
    "                \n",
    "            \n",
    "            # 2. If the SNP is within 1000 bp of the nearest sequence end [DIST], record as N. \n",
    "            if dist <= 1000:\n",
    "                alt = 'N'\n",
    "                alt_distance += 1\n",
    "                \n",
    "                \n",
    "            # 3. ignore insertions \n",
    "            if ref != '.':\n",
    "                \n",
    "                ##### SAVE THE ALT ALLELE\n",
    "                sample_index = column_names.index(col_name)\n",
    "                position_index = positions.index(pos)\n",
    "                #\n",
    "                rows[sample_index][position_index] = alt\n",
    "                alt_alt += 1\n",
    "\n",
    "            \n",
    "    \n",
    "\n",
    "       \n",
    "    print('marked N due to deletion: ', alt_delete)\n",
    "    print('marked N due to within 1000 bp of sequence end: ',alt_distance)\n",
    "    print('alt allele recorded: ',alt_alt)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66393.94267515924\n",
      "159633.46496815287\n"
     ]
    }
   ],
   "source": [
    "numN =[]\n",
    "numNone=[]\n",
    "\n",
    "for i in rows:\n",
    "    numN.append(i.count(\"N\")) \n",
    "    numNone.append(i.count(None))\n",
    "                \n",
    "print(sum(numN)/len(numN)) # mean 66393 / 246625 = 27% of SNP positions are missing data, including no alignment or SNP called in less certain region\n",
    "print(sum(numNone)/len(numNone)) # mean 159633 / 246625 = 64% of positions have data and the REF allele (To be filled in next cell)\n",
    "# This means the remaining 20599 (8%) of SNPs have ALT allele\n",
    "# These nubmers are consistent with previous iteration\n",
    "\n",
    "\n",
    "##@@ Save 'rows' again now that the alt alleles are filled in \n",
    "json_string = json.dumps(rows)\n",
    "with open(path+\"rows_vALT_032525.json\", \"w\") as f:\n",
    "    f.write(json_string)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157\n",
      "246625\n",
      "157\n",
      "ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425\n"
     ]
    }
   ],
   "source": [
    "# Jump in midstream\n",
    "\n",
    "import json\n",
    "import csv\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "\n",
    "path=\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/\"\n",
    "with open(path+'/rows_vALT_032525.json', 'r') as f:\n",
    "    rows = json.load(f)\n",
    "print(len(rows))   \n",
    "\n",
    "with open(path+'/position_list_032525.json', 'r') as f:\n",
    "    positions = json.load(f)  \n",
    "print(len(positions)) \n",
    "\n",
    "coordsfiles = []\n",
    "files = os.listdir(path)\n",
    "for i in files:\n",
    "    if i[-7:] == '.coords':\n",
    "        coordsfiles.append(i) \n",
    "\n",
    "column_names = [] \n",
    "for i in coordsfiles:\n",
    "    i2 = i.split('ATCC_v_')[1]\n",
    "    i3 = i2.split('.coords')[0]\n",
    "    column_names.append(i3)\n",
    "    \n",
    "\n",
    "print(len(column_names))\n",
    "print(column_names[99])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "247 K positions. 27% are missing, so only 180K positions. Of those 159581 or 89% have the ref ATCC allele. The remaining 11% of positions with data differ from the ref."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2089645\n",
      "[1200, 1371, 1695, 1756, 1799]\n",
      "['G', 'C', 'A', 'A', 'G']\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "############################################## \n",
    "###  Replace 'None' with the REF allele (speedy)  ###\n",
    "############################################## \n",
    "\n",
    "### GET THE GENOME (since I failed to record this earlier when going through the SNP files)\n",
    "# in the form of a list x base pairs long \n",
    "\n",
    "genomelist = []\n",
    "\n",
    "with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/ncbi_genomes/ATCC-15703_lab-strain_GCA_000010425.1.fna\",\"r\") as csvfile:\n",
    "    reader = csv.reader(csvfile, delimiter='\\t') \n",
    "    header=next(reader)\n",
    "    for row in reader:\n",
    "\n",
    "        genomelist.append(row[0])\n",
    "        \n",
    "genomestring = \"\".join(genomelist)\n",
    "print(len(genomestring))\n",
    "\n",
    "\n",
    "##@@ CHECK ABILITY TO PULL REF ALLELES \n",
    "# print(genomestring[1200-1]) # snps.txt file says posisiton 1200 is a G but python has 0 indexing\n",
    "# print(genomestring[1371-1]) # these all work; carry on\n",
    "# print(genomestring[1695-1])\n",
    "# print(genomestring[1756-1])\n",
    "# print(genomestring[1799-1])\n",
    "# print(genomestring[36031-1])\n",
    "\n",
    "\n",
    "##@@  PULL REF ALLELES \n",
    "ref_alleles = []\n",
    "ref_positions = []  ### record positions as integers \n",
    "\n",
    "for k in range(len(positions)):\n",
    "    ref_positions.append(int(positions[k]))\n",
    "    \n",
    "\n",
    "for s in ref_positions:\n",
    "    ref_alleles.append(genomestring[s-1])\n",
    "    \n",
    "print(ref_positions[0:5])\n",
    "print(ref_alleles[0:5])\n",
    "\n",
    "\n",
    "##@@ FILL IN REF ALLELES \n",
    "\n",
    "count = -1\n",
    "for sample_list in rows:\n",
    "    count+=1\n",
    "\n",
    "\n",
    "    for i in range(len(sample_list)):  ## looping through the SNV positions \n",
    "        if sample_list[i] == None: \n",
    "            # The index here is the same as the index of ref_allele\n",
    "            sample_list[i] = ref_alleles[i]\n",
    "            \n",
    "\n",
    "\n",
    "print('Done!')            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        LMG-10502_type_GCA_028743295 Mongolia-MGYG000240385  \\\n",
      "1200                               G                      A   \n",
      "1371                               C                      T   \n",
      "1695                               A                      G   \n",
      "1756                               A                      N   \n",
      "1799                               G                      A   \n",
      "...                              ...                    ...   \n",
      "2062761                            C                      C   \n",
      "2062766                            T                      T   \n",
      "2062768                            A                      A   \n",
      "2063055                            A                      A   \n",
      "2076218                            C                      C   \n",
      "\n",
      "        AM66-51ACA_human-feces_China_GCA_027667065  \\\n",
      "1200                                             A   \n",
      "1371                                             T   \n",
      "1695                                             G   \n",
      "1756                                             N   \n",
      "1799                                             A   \n",
      "...                                            ...   \n",
      "2062761                                          C   \n",
      "2062766                                          T   \n",
      "2062768                                          A   \n",
      "2063055                                          N   \n",
      "2076218                                          C   \n",
      "\n",
      "        1001713B-A6_human-feces_NewYork_GCA_015558745 Mongolia-MGYG000100616  \\\n",
      "1200                                                A                      A   \n",
      "1371                                                T                      T   \n",
      "1695                                                G                      G   \n",
      "1756                                                N                      N   \n",
      "1799                                                A                      A   \n",
      "...                                               ...                    ...   \n",
      "2062761                                             N                      N   \n",
      "2062766                                             N                      N   \n",
      "2062768                                             A                      A   \n",
      "2063055                                             A                      A   \n",
      "2076218                                             C                      C   \n",
      "\n",
      "        Mongolia-MGYG000012063 Fiji-MGYG000099035  \\\n",
      "1200                         A                  A   \n",
      "1371                         N                  T   \n",
      "1695                         N                  G   \n",
      "1756                         N                  N   \n",
      "1799                         N                  A   \n",
      "...                        ...                ...   \n",
      "2062761                      C                  C   \n",
      "2062766                      T                  T   \n",
      "2062768                      A                  A   \n",
      "2063055                      A                  N   \n",
      "2076218                      C                  C   \n",
      "\n",
      "        Bifidobacterium-adolescentis_AK_SK_36_Boto UnitedStates-MGYG000241712  \\\n",
      "1200                                             A                          A   \n",
      "1371                                             T                          T   \n",
      "1695                                             G                          G   \n",
      "1756                                             N                          N   \n",
      "1799                                             A                          A   \n",
      "...                                            ...                        ...   \n",
      "2062761                                          N                          N   \n",
      "2062766                                          N                          N   \n",
      "2062768                                          N                          N   \n",
      "2063055                                          N                          A   \n",
      "2076218                                          C                          C   \n",
      "\n",
      "        Madagascar-MGYG000120594  ...  \\\n",
      "1200                           A  ...   \n",
      "1371                           T  ...   \n",
      "1695                           G  ...   \n",
      "1756                           N  ...   \n",
      "1799                           A  ...   \n",
      "...                          ...  ...   \n",
      "2062761                        N  ...   \n",
      "2062766                        N  ...   \n",
      "2062768                        N  ...   \n",
      "2063055                        N  ...   \n",
      "2076218                        C  ...   \n",
      "\n",
      "        1001275B-D1_human-feces_NewYork_GCA_015558565  \\\n",
      "1200                                                A   \n",
      "1371                                                T   \n",
      "1695                                                G   \n",
      "1756                                                N   \n",
      "1799                                                A   \n",
      "...                                               ...   \n",
      "2062761                                             C   \n",
      "2062766                                             T   \n",
      "2062768                                             A   \n",
      "2063055                                             N   \n",
      "2076218                                             C   \n",
      "\n",
      "        CLA-AP-H11_human-feces_Germany_GCA_040092935 Mongolia-MGYG000175950  \\\n",
      "1200                                               A                      A   \n",
      "1371                                               T                      T   \n",
      "1695                                               G                      G   \n",
      "1756                                               N                      N   \n",
      "1799                                               A                      A   \n",
      "...                                              ...                    ...   \n",
      "2062761                                            C                      C   \n",
      "2062766                                            T                      T   \n",
      "2062768                                            A                      A   \n",
      "2063055                                            N                      N   \n",
      "2076218                                            C                      C   \n",
      "\n",
      "        ElSalvador-MGYG000132829 Bifidobacterium-adolescentis_RS12_B_Warli  \\\n",
      "1200                           G                                         N   \n",
      "1371                           C                                         T   \n",
      "1695                           G                                         G   \n",
      "1756                           N                                         N   \n",
      "1799                           A                                         A   \n",
      "...                          ...                                       ...   \n",
      "2062761                        C                                         C   \n",
      "2062766                        T                                         T   \n",
      "2062768                        A                                         A   \n",
      "2063055                        N                                         N   \n",
      "2076218                        C                                         C   \n",
      "\n",
      "        CLA-SR-H001_human-feces_Germany_GCA_040097245  \\\n",
      "1200                                                A   \n",
      "1371                                                T   \n",
      "1695                                                G   \n",
      "1756                                                N   \n",
      "1799                                                A   \n",
      "...                                               ...   \n",
      "2062761                                             C   \n",
      "2062766                                             T   \n",
      "2062768                                             A   \n",
      "2063055                                             N   \n",
      "2076218                                             C   \n",
      "\n",
      "        AM64-16A_China_GCA_027667565 UnitedStates-MGYG000225709  \\\n",
      "1200                               A                          A   \n",
      "1371                               C                          T   \n",
      "1695                               G                          G   \n",
      "1756                               N                          N   \n",
      "1799                               G                          A   \n",
      "...                              ...                        ...   \n",
      "2062761                            C                          C   \n",
      "2062766                            T                          T   \n",
      "2062768                            A                          A   \n",
      "2063055                            N                          N   \n",
      "2076218                            C                          N   \n",
      "\n",
      "        REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ  \\\n",
      "1200                                                     A                                                       \n",
      "1371                                                     T                                                       \n",
      "1695                                                     G                                                       \n",
      "1756                                                     N                                                       \n",
      "1799                                                     A                                                       \n",
      "...                                                    ...                                                       \n",
      "2062761                                                  N                                                       \n",
      "2062766                                                  N                                                       \n",
      "2062768                                                  N                                                       \n",
      "2063055                                                  N                                                       \n",
      "2076218                                                  C                                                       \n",
      "\n",
      "        Bifidobacterium-adolescentis_AK_SK_25_Purigpa  \n",
      "1200                                                A  \n",
      "1371                                                T  \n",
      "1695                                                G  \n",
      "1756                                                N  \n",
      "1799                                                A  \n",
      "...                                               ...  \n",
      "2062761                                             N  \n",
      "2062766                                             N  \n",
      "2062768                                             N  \n",
      "2063055                                             N  \n",
      "2076218                                             C  \n",
      "\n",
      "[246625 rows x 157 columns]\n",
      "Index(['LMG-10502_type_GCA_028743295', 'Mongolia-MGYG000240385',\n",
      "       'AM66-51ACA_human-feces_China_GCA_027667065',\n",
      "       '1001713B-A6_human-feces_NewYork_GCA_015558745',\n",
      "       'Mongolia-MGYG000100616', 'Mongolia-MGYG000012063',\n",
      "       'Fiji-MGYG000099035', 'Bifidobacterium-adolescentis_AK_SK_36_Boto',\n",
      "       'UnitedStates-MGYG000241712', 'Madagascar-MGYG000120594',\n",
      "       ...\n",
      "       '1001275B-D1_human-feces_NewYork_GCA_015558565',\n",
      "       'CLA-AP-H11_human-feces_Germany_GCA_040092935',\n",
      "       'Mongolia-MGYG000175950', 'ElSalvador-MGYG000132829',\n",
      "       'Bifidobacterium-adolescentis_RS12_B_Warli',\n",
      "       'CLA-SR-H001_human-feces_Germany_GCA_040097245',\n",
      "       'AM64-16A_China_GCA_027667565', 'UnitedStates-MGYG000225709',\n",
      "       'REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ',\n",
      "       'Bifidobacterium-adolescentis_AK_SK_25_Purigpa'],\n",
      "      dtype='object', length=157)\n"
     ]
    }
   ],
   "source": [
    "###########################################################################\n",
    "#######        FINALLY, BUILD THE DATA FRAME AND LOOK AT IT\n",
    "###########################################################################\n",
    "\n",
    "# it's originally the orientation I don't want\n",
    "df_transposed = pd.DataFrame(rows,columns=positions,index=column_names)\n",
    "\n",
    "# so transpose it \n",
    "df = df_transposed.T \n",
    "print(df)\n",
    "#df.to_csv(path+'SNV_table_032625.csv', index=True) # SNV_table_102124.csv was  overwritten, whoops\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1-11_infant-feces_China_GCA_003030905\n",
      "1001099B-F6_human-feces_NewYork_GCA_015554265\n",
      "1001262B-human-feces_NewYork_GCA_015557685\n",
      "1001270B-E2_human-feces_NewYork_GCA_015548665\n",
      "1001271B-A4_human-feces_NewYork_GCA_015548755\n",
      "1001275B-D1_human-feces_NewYork_GCA_015558565\n",
      "1001283B1-G1_human-feces_NewYork_GCA_015557595\n",
      "1001713B-A6_human-feces_NewYork_GCA_015558745\n",
      "150_human-feces_Russia_GCA_001010915\n",
      "2789STDY5834850_human-feces_UnitedKingdom_GCA_001406215\n",
      "2789ST_UK_GCA_001406455\n",
      "AF02-8_human-feces_China_GCA_027688555\n",
      "AF14-56_human-feces_China_GCA_003465205\n",
      "AF92-02A_human-feces_China_GCA_027664845\n",
      "AM110-65_human-feces_China_GCA_027674825\n",
      "AM14-37_China_GCA_003472095\n",
      "AM64-16A_China_GCA_027667565\n",
      "AM66-51ACA_human-feces_China_GCA_027667065\n",
      "ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425\n",
      "BB23_human-feces_China_GCA_000817995\n",
      "Bifidobacterium-adolescentis_AK_SK_11_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_12_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_13_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_1_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_25_Purigpa\n",
      "Bifidobacterium-adolescentis_AK_SK_2_Purigpa\n",
      "Bifidobacterium-adolescentis_AK_SK_32_Balti\n",
      "Bifidobacterium-adolescentis_AK_SK_33_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_35_Balti\n",
      "Bifidobacterium-adolescentis_AK_SK_36_Boto\n",
      "Bifidobacterium-adolescentis_AK_SK_37_Brokpa\n",
      "Bifidobacterium-adolescentis_AK_SK_44_Balti\n",
      "Bifidobacterium-adolescentis_AK_SK_45_Purigpa\n",
      "Bifidobacterium-adolescentis_AK_SK_48_Purigpa\n",
      "Bifidobacterium-adolescentis_AK_SK_8_Balti\n",
      "Bifidobacterium-adolescentis_AK_SK_9_Boto\n",
      "Bifidobacterium-adolescentis_RS12_A_Warli\n",
      "Bifidobacterium-adolescentis_RS12_B_Warli\n",
      "Bifidobacterium-adolescentis_RS5_A_Warli\n",
      "CLA-AP-H11_human-feces_Germany_GCA_040092935\n",
      "CLA-SR-H001_human-feces_Germany_GCA_040097245\n",
      "China-MGYG000047373\n",
      "China-MGYG000093441\n",
      "China-MGYG000111016\n",
      "China-MGYG000117766\n",
      "China-MGYG000170701\n",
      "China-MGYG000170766\n",
      "China-MGYG000175018\n",
      "China-MGYG000225533\n",
      "China-MGYG000226008\n",
      "China-MGYG000287017\n",
      "D31t1-E3_human-feces_NewYork_GCA_015549865\n",
      "D46t1-C9_human-feces_NewYork_GCA_015558575\n",
      "D52t1-B8_human-feces_NewYork_GCA_015558415\n",
      "D53t1_human-feces_NewYork_GCA_015552825\n",
      "D59t2-181005-G7_human-feces_NewYork_GCA_015555595\n",
      "ElSalvador-MGYG000032473\n",
      "ElSalvador-MGYG000037768\n",
      "ElSalvador-MGYG000099659\n",
      "ElSalvador-MGYG000124287\n",
      "ElSalvador-MGYG000132829\n",
      "ElSalvador-MGYG000209388\n",
      "ElSalvador-MGYG000210883\n",
      "ElSalvador-MGYG000228997\n",
      "ElSalvador-MGYG000284637\n",
      "ElSalvador-MGYG000285386\n",
      "Fiji-MGYG000031842\n",
      "Fiji-MGYG000076435\n",
      "Fiji-MGYG000098158\n",
      "Fiji-MGYG000099035\n",
      "Fiji-MGYG000104252\n",
      "Fiji-MGYG000124014\n",
      "Fiji-MGYG000199136\n",
      "Fiji-MGYG000248641\n",
      "Fiji-MGYG000255607\n",
      "Fiji-MGYG000285608\n",
      "H37_UK_GCA_001406735\n",
      "H411_UK_GCA_001406215\n",
      "JCM-15918_sbmtMongolia_GCA_000771705\n",
      "JCM15918_type-strain_NA_GCA_000741415\n",
      "Km4_human-feces_Russia_GCA_001756865\n",
      "L2-32_human-feces_NA_GCA_000154085\n",
      "LMG-10502_type_GCA_028743295\n",
      "LMG-10733_type_GCA_002108135\n",
      "MCC257_human-feces_Ireland_GCA_018785715\n",
      "METABAT215_SUBJECTMAPPING_SCAFFOLDS_1500_MERGED_K77_METASPADES__AK_SK_5\n",
      "Madagascar-MGYG000012482\n",
      "Madagascar-MGYG000025145\n",
      "Madagascar-MGYG000091572\n",
      "Madagascar-MGYG000120594\n",
      "Mongolia-MGYG000008798\n",
      "Mongolia-MGYG000012063\n",
      "Mongolia-MGYG000021867\n",
      "Mongolia-MGYG000048473\n",
      "Mongolia-MGYG000050898\n",
      "Mongolia-MGYG000056268\n",
      "Mongolia-MGYG000059995\n",
      "Mongolia-MGYG000060881\n",
      "Mongolia-MGYG000066585\n",
      "Mongolia-MGYG000073223\n",
      "Mongolia-MGYG000100616\n",
      "Mongolia-MGYG000116659\n",
      "Mongolia-MGYG000117819\n",
      "Mongolia-MGYG000131768\n",
      "Mongolia-MGYG000158456\n",
      "Mongolia-MGYG000175950\n",
      "Mongolia-MGYG000185946\n",
      "Mongolia-MGYG000212973\n",
      "Mongolia-MGYG000240385\n",
      "Mongolia-MGYG000280676\n",
      "NCTC11814_type_GCA_900445615\n",
      "REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_21_7002\n",
      "REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_24_1038\n",
      "REFINED_METABAT215_SUBJECT_CONTIGS_1500_ASSEMBLY_K77_MERGED__Pilot_MoBio_Fiber_C_4_7003\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_A_10_1818\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_E_24_1866\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_P_19_2149\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_B_17_2661\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_D_23_2656\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_21_2500\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-I-L_M_23_2501\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_17_RAU1026YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_B_21_THA1063YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_3_THA1067YZ_1\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_5_THA1074JZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_11_RAU1009YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_23_THA0068JZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_F_5_RAU0022YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_H_3_THA1060YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_15_THA1075YZ\n",
      "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_J_7_THA0058AZ\n",
      "UnitedKingdom-MGYG000036973\n",
      "UnitedKingdom-MGYG000044575\n",
      "UnitedKingdom-MGYG000044941\n",
      "UnitedKingdom-MGYG000083139\n",
      "UnitedKingdom-MGYG000114505\n",
      "UnitedKingdom-MGYG000124758\n",
      "UnitedKingdom-MGYG000146104\n",
      "UnitedKingdom-MGYG000196958\n",
      "UnitedKingdom-MGYG000205281\n",
      "UnitedKingdom-MGYG000287127\n",
      "UnitedStates-MGYG000092376\n",
      "UnitedStates-MGYG000092433\n",
      "UnitedStates-MGYG000101498\n",
      "UnitedStates-MGYG000114143\n",
      "UnitedStates-MGYG000136926\n",
      "UnitedStates-MGYG000138549\n",
      "UnitedStates-MGYG000147020\n",
      "UnitedStates-MGYG000225709\n",
      "UnitedStates-MGYG000241712\n",
      "UnitedStates-MGYG000283239\n",
      "VKPM-Ac-1245_human-feces_Russia_GCA_016069975\n",
      "ZJ2-centenarian-feces_China_GCA_009832825\n",
      "ca0067_human-feces_CambridgeMA_GCA_004167585\n",
      "iVS-1_human-feces_Nebraska_GCA_029855065\n"
     ]
    }
   ],
   "source": [
    "column_names = df.columns.tolist()\n",
    "column_names.sort()\n",
    "for i in column_names:\n",
    "    \n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1200</th>\n",
       "      <th>1371</th>\n",
       "      <th>1695</th>\n",
       "      <th>1756</th>\n",
       "      <th>1799</th>\n",
       "      <th>1802</th>\n",
       "      <th>1805</th>\n",
       "      <th>1816</th>\n",
       "      <th>1817</th>\n",
       "      <th>1827</th>\n",
       "      <th>...</th>\n",
       "      <th>1984442</th>\n",
       "      <th>1998225</th>\n",
       "      <th>1999559</th>\n",
       "      <th>2044350</th>\n",
       "      <th>2062734</th>\n",
       "      <th>2062761</th>\n",
       "      <th>2062766</th>\n",
       "      <th>2062768</th>\n",
       "      <th>2063055</th>\n",
       "      <th>2076218</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Mongolia-MGYG000240385</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM66-51ACA_human-feces_China_GCA_027667065</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001713B-A6_human-feces_NewYork_GCA_015558745</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mongolia-MGYG000100616</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mongolia-MGYG000012063</th>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLA-SR-H001_human-feces_Germany_GCA_040097245</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM64-16A_China_GCA_027667565</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UnitedStates-MGYG000225709</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bifidobacterium-adolescentis_AK_SK_25_Purigpa</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132 rows  246625 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   1200    1371    1695     \\\n",
       "Mongolia-MGYG000240385                                   A       T       G   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               A       T       G   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            A       T       G   \n",
       "Mongolia-MGYG000100616                                   A       T       G   \n",
       "Mongolia-MGYG000012063                                   A       N       N   \n",
       "...                                                    ...     ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            A       T       G   \n",
       "AM64-16A_China_GCA_027667565                             A       C       G   \n",
       "UnitedStates-MGYG000225709                               A       T       G   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       A       T       G   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            A       T       G   \n",
       "\n",
       "                                                   1756    1799    1802     \\\n",
       "Mongolia-MGYG000240385                                   N       A       T   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               N       A       T   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            N       A       T   \n",
       "Mongolia-MGYG000100616                                   N       A       T   \n",
       "Mongolia-MGYG000012063                                   N       N       N   \n",
       "...                                                    ...     ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            N       A       T   \n",
       "AM64-16A_China_GCA_027667565                             N       G       A   \n",
       "UnitedStates-MGYG000225709                               N       A       T   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       N       A       T   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            N       A       T   \n",
       "\n",
       "                                                   1805    1816    1817     \\\n",
       "Mongolia-MGYG000240385                                   C       G       A   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               C       G       A   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            C       G       A   \n",
       "Mongolia-MGYG000100616                                   C       G       A   \n",
       "Mongolia-MGYG000012063                                   N       N       N   \n",
       "...                                                    ...     ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            C       G       A   \n",
       "AM64-16A_China_GCA_027667565                             T       A       G   \n",
       "UnitedStates-MGYG000225709                               C       G       A   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       C       G       A   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            C       G       A   \n",
       "\n",
       "                                                   1827     ... 1984442  \\\n",
       "Mongolia-MGYG000240385                                   G  ...       N   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               G  ...       G   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            G  ...       G   \n",
       "Mongolia-MGYG000100616                                   G  ...       G   \n",
       "Mongolia-MGYG000012063                                   N  ...       G   \n",
       "...                                                    ...  ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            G  ...       G   \n",
       "AM64-16A_China_GCA_027667565                             G  ...       G   \n",
       "UnitedStates-MGYG000225709                               G  ...       G   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       G  ...       G   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            G  ...       N   \n",
       "\n",
       "                                                   1998225 1999559 2044350  \\\n",
       "Mongolia-MGYG000240385                                   A       G       C   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               A       G       N   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            A       G       C   \n",
       "Mongolia-MGYG000100616                                   A       G       C   \n",
       "Mongolia-MGYG000012063                                   A       G       C   \n",
       "...                                                    ...     ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            A       G       N   \n",
       "AM64-16A_China_GCA_027667565                             A       G       N   \n",
       "UnitedStates-MGYG000225709                               A       G       C   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       N       G       N   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            N       N       C   \n",
       "\n",
       "                                                   2062734 2062761 2062766  \\\n",
       "Mongolia-MGYG000240385                                   C       C       T   \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               C       C       T   \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            N       N       N   \n",
       "Mongolia-MGYG000100616                                   N       N       N   \n",
       "Mongolia-MGYG000012063                                   C       C       T   \n",
       "...                                                    ...     ...     ...   \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            C       C       T   \n",
       "AM64-16A_China_GCA_027667565                             C       C       T   \n",
       "UnitedStates-MGYG000225709                               C       C       T   \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       N       N       N   \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            N       N       N   \n",
       "\n",
       "                                                   2062768 2063055 2076218  \n",
       "Mongolia-MGYG000240385                                   A       A       C  \n",
       "AM66-51ACA_human-feces_China_GCA_027667065               A       N       C  \n",
       "1001713B-A6_human-feces_NewYork_GCA_015558745            A       A       C  \n",
       "Mongolia-MGYG000100616                                   A       A       C  \n",
       "Mongolia-MGYG000012063                                   A       A       C  \n",
       "...                                                    ...     ...     ...  \n",
       "CLA-SR-H001_human-feces_Germany_GCA_040097245            A       N       C  \n",
       "AM64-16A_China_GCA_027667565                             A       N       C  \n",
       "UnitedStates-MGYG000225709                               A       N       N  \n",
       "REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_...       N       N       C  \n",
       "Bifidobacterium-adolescentis_AK_SK_25_Purigpa            N       N       C  \n",
       "\n",
       "[132 rows x 246625 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Remove certain samples here\n",
    "# banana\n",
    "\n",
    "\n",
    "\n",
    "# remove = ['Madagascar-MGYG000012482', 'Madagascar-MGYG000025145', 'Madagascar-MGYG000091572', 'Madagascar-MGYG000120594',\n",
    "# 'Fiji-MGYG000031842', 'Fiji-MGYG000076435', 'Fiji-MGYG000098158', 'Fiji-MGYG000099035', 'Fiji-MGYG000104252', 'Fiji-MGYG000124014', 'Fiji-MGYG000199136', 'Fiji-MGYG000248641', 'Fiji-MGYG000255607', 'Fiji-MGYG000285608',\n",
    "# \"ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425\",\"JCM-15918_sbmtMongolia_GCA_000771705\",\n",
    "# \"JCM15918_type-strain_NA_GCA_000741415\",\"L2-32_human-feces_NA_GCA_000154085\",\"LMG-10502_type_GCA_028743295\",\"LMG-10733_type_GCA_002108135\",\"NCTC11814_type_GCA_900445615\",\n",
    "#     \"Km4_human-feces_Russia_GCA_001756865\",\"1-11_infant-feces_China_GCA_003030905\",\"150_human-feces_Russia_GCA_001010915\",\"VKPM-Ac-1245_human-feces_Russia_GCA_016069975\"]\n",
    "\n",
    "# print(len(remove))\n",
    "\n",
    "\n",
    "# df2 = df.drop(remove, axis=1)\n",
    "\n",
    "\n",
    "# great, but filtering used 'df_transposed'\n",
    "df_transposed_2 = df2.T\n",
    "\n",
    "\n",
    "df_transposed_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 4407, 27545, 28790, 31453, 31740, 31926, 32813, 33473, 33827, 33900, 34030, 34511, 34839, 35068, 35277, 35408, 35627, 36016, 36329, 36550, 36760, 36760, 36879, 37656, 37937, 38232, 38293, 38374, 38560, 38707, 39408, 39610, 40074, 40486, 41375, 41525, 41627, 41719, 42114, 42214, 42536, 42971, 43206, 43433, 43798, 45930, 45997, 46308, 46362, 46606, 46660, 46679, 47166, 47737, 49962, 50268, 50954, 52149, 52747, 53458, 55524, 55724, 56067, 56736, 57240, 58559, 59514, 59730, 60016, 61240, 63149, 63354, 63962, 64247, 64922, 65450, 65820, 66367, 67341, 67824, 67825, 67983, 68397, 68788, 69246, 69734, 70107, 70267, 71952, 72170, 72324, 72729, 75352, 76097, 76759, 76999, 77271, 78518, 79764, 80342, 81133, 81458, 81802, 81810, 83187, 83327, 83464, 83537, 83723, 84992, 85596, 85732, 85802, 86511, 86596, 87198, 87257, 87296, 88106, 88353, 88527, 88609, 89015, 89180, 89388, 90700, 91007, 91973, 93805, 94239, 94341, 95254, 96232, 97589, 97690, 97703, 97869, 99114, 99186, 100574, 100622, 102209, 103631, 108848, 110348, 114934, 117040, 117188, 118432, 118803, 121851, 129988, 134273, 134672, 139959]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "### View how many SNVs are missing (out of 246625 before filtering)\n",
    "column_counts = df.apply(lambda col: col.value_counts().get('N', 0))\n",
    "cc = list(column_counts)\n",
    "print(sorted(cc))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting SNVs:  246625\n",
      "Ending SNVs:  108799\n",
      "Not biallelic:  64362\n",
      "Many N:  16658\n",
      "Single sample:  56806\n"
     ]
    }
   ],
   "source": [
    "# correct filtering \n",
    "\n",
    "\n",
    "\n",
    "#### FILTERING POSITIONS \n",
    "\n",
    "\n",
    "    # 1     # Positions that are not actually variable, i.e., only REF or N. \n",
    "            # These may be too close to ends of chromosomes to confidently call variants.\n",
    "        \n",
    "    # 2     # Where N is very high, e.g. pos 1756. Could be  a common deletion, since deletions were recorded as Ns. (\n",
    "\n",
    "    # 3     # Multi-allelic positions, to help control for mis-alignment and paralogs \n",
    "    \n",
    "    # 4     # Positions where only one sample has the ALT allele, because these will not be phylogenetically informative\n",
    "\n",
    "# Actually, I think it will be easier to  convert it back to lists, \n",
    "# where this time, each list is a SNP that is 157 items long \n",
    "    \n",
    "    \n",
    "## In practice, it will convert back to a dict, where the key is the SNP position  \n",
    "\n",
    "#column_dict = df_transposed.to_dict('list') # N =157\n",
    "column_dict = df_transposed_2.to_dict('list') # N=133\n",
    "kys = list(column_dict.keys()) # the key is a numeric SNV position\n",
    "\n",
    "print(\"Starting SNVs: \",len(column_dict))\n",
    "\n",
    "## Prepare to filter\n",
    "filtered_dict = {}\n",
    "bad_dict = {}\n",
    "not_biallelic = 0\n",
    "manyN = 0\n",
    "singlesample = 0\n",
    "\n",
    "\n",
    "for snv in kys:\n",
    "    \n",
    "    snvpass = False\n",
    "    \n",
    "    genotypes = column_dict[snv]\n",
    "    \n",
    "    # count each base\n",
    "    A = genotypes.count('A')\n",
    "    T = genotypes.count('T')\n",
    "    C = genotypes.count('C')\n",
    "    G = genotypes.count('G')\n",
    "    N = genotypes.count('N')\n",
    "\n",
    "    \n",
    "    # Require variation, i.e., no more than 3 of the base sums can be 0. \n",
    "    # At the same time, exclude multi-allelic sites.\n",
    "    basesum = [A,T,C,G]\n",
    "    if basesum.count(0) == 2:\n",
    "        \n",
    "        # Require that at least half the samples have data \n",
    "        if N <= 80:\n",
    "                 \n",
    "            # Finally, require that more than one sample have the alt allele \n",
    "            basesum.remove(0)\n",
    "            basesum.remove(0)\n",
    "            if min(basesum) > 1:\n",
    "                snvpass = True\n",
    "            else: \n",
    "                singlesample += 1                   \n",
    "        else:\n",
    "            manyN += 1          \n",
    "    else:\n",
    "        not_biallelic += 1\n",
    "    \n",
    "    \n",
    "    ### If it's a good snv, save in a new dict\n",
    "    if snvpass == True:\n",
    "        filtered_dict[snv] = genotypes\n",
    "    else:\n",
    "        bad_dict[snv] = genotypes\n",
    "\n",
    "\n",
    "\n",
    "print(\"Ending SNVs: \",len(filtered_dict))\n",
    "print(\"Not biallelic: \",not_biallelic)\n",
    "print(\"Many N: \",manyN)\n",
    "print(\"Single sample: \", singlesample)\n",
    "\n",
    " \n",
    "        \n",
    "# N=157\n",
    "# Starting SNVs:  246625\n",
    "# Ending SNVs:  113305\n",
    "# Not biallelic:  46650\n",
    "# Many N:  25826\n",
    "# Single sample:  60844\n",
    "\n",
    "# N=132 \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Length mismatch: Expected axis has 132 elements, new values have 157 elements",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1018533/3997378117.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_dict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcolumn_names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mdfT\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   5498\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5499\u001b[0m             \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5500\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5501\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5502\u001b[0m             \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/_libs/properties.pyx\u001b[0m in \u001b[0;36mpandas._libs.properties.AxisProperty.__set__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_set_axis\u001b[0;34m(self, axis, labels)\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m         \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 766\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    767\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    768\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mset_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m    214\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mset_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    215\u001b[0m         \u001b[0;31m# Caller is responsible for ensuring we have an Index object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 216\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    217\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    218\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/internals/base.py\u001b[0m in \u001b[0;36m_validate_set_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mnew_len\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mold_len\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m             raise ValueError(\n\u001b[0m\u001b[1;32m     58\u001b[0m                 \u001b[0;34mf\"Length mismatch: Expected axis has {old_len} elements, new \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m                 \u001b[0;34mf\"values have {new_len} elements\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Length mismatch: Expected axis has 132 elements, new values have 157 elements"
     ]
    }
   ],
   "source": [
    "### WRITE THE FILTERED_DICT TO A SNV FILE \n",
    "\n",
    "df = pd.DataFrame(filtered_dict)\n",
    "df.index = column_names ## WRONG when N = 133\n",
    "\n",
    "dfT = df.T \n",
    "print(dfT)\n",
    "#dfT.to_csv(path+'SNV_table_FILT_032526.csv', index=True)\n",
    "\n",
    "\n",
    "# correctly, columns are samples and rows are positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert this back to a dataframe, transpose, and back to lists \n",
    "\n",
    "df_filt_transposed = pd.DataFrame.from_dict(filtered_dict)\n",
    "df_filt_transposed.index = column_names\n",
    "#df_filt_transposed\n",
    "\n",
    "df_filt = df_filt_transposed.T \n",
    "df_filt\n",
    "\n",
    "final_dict = df_filt.to_dict('list')\n",
    "final_keys = list(final_dict.keys()) # the key is a numeric SNV position\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(df_filt)\n",
    "# final_dict[final_keys[0]]\n",
    "# final_keys[0:5]\n",
    "# print(column_names[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "### NOW SHORTEN NAMES FOR TREE \n",
    "\n",
    "## For shortening names \n",
    "name_dict ={} # key is long name, value is short name \n",
    "\n",
    "with open(path+\"name_key_raxml_032625.txt\",\"r\") as csvfile:\n",
    "    reader = csv.reader(csvfile, delimiter='\\t') \n",
    "    header=next(reader)\n",
    "    for row in reader:\n",
    "        if row[0] != '':\n",
    "            longname = row[0].split('.')[0]\n",
    "            shortname = row[1]\n",
    "\n",
    "            name_dict[longname] = shortname\n",
    "\n",
    "#print(len(name_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "132\n",
      "157\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'final_keys' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1018533/574851756.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mdropped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0mwritten\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfinal_keys\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfinal_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'final_keys' is not defined"
     ]
    }
   ],
   "source": [
    "###########################################################################\n",
    "#######       WRITE TO FILE FOR TREE \n",
    "###########################################################################\n",
    "\n",
    "\n",
    "######## RAXML INPUT FORMAT \n",
    "\n",
    "#181     # N samples \n",
    "#143468  # N sites\n",
    "#AK_SG_16 GTTTGGGGTGTGAAGGTTAGAAGATCGTCCGGGCG...\n",
    "\n",
    "\n",
    "path = \"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/March2025/\"\n",
    "\n",
    "#outfile = open(path+\"Bif-adol_157_113305_raxml_032625.txt\",\"w\")\n",
    "outfile = open(path+\"Bif-adol_132_113305_raxml_032725.txt\",\"w\")\n",
    "outfile.write('132'+'\\n'+'113305'+'\\n')\n",
    "\n",
    "dropped = []\n",
    "written = []\n",
    "for s in final_keys:\n",
    "    \n",
    "    data = final_dict[s]\n",
    "    name = name_dict[s]\n",
    "\n",
    "    if name in keep_short:\n",
    "    \n",
    "        outfile.write(name+' ')  \n",
    "        written.append(name)\n",
    "        for a in data:\n",
    "            outfile.write(a)\n",
    "        outfile.write('\\n')\n",
    "    else:\n",
    "        print(s)\n",
    "        dropped.append(name)\n",
    "\n",
    "outfile.close()\n",
    "\n",
    "print(len(dropped))\n",
    "\n",
    "\n",
    "# for i in keep_short:\n",
    "#     if i not in written:\n",
    "#         print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATCC_type', 'China_infI', 'Fiji_1042M', 'Fiji_1240M', 'Fiji_1991M', 'Fiji_2486M', 'Fiji_2556M', 'Fiji_2856M', 'Fiji_3184M', 'Fiji_7643M', 'Fiji_9815M', 'Fiji_9903M', 'JCM_type', 'Korea_type', 'LMG10502type', 'LMG10733type', 'Madagas_120M', 'Madagas_124M', 'Madagas_251M', 'Madagas_915M', 'NCTC_type', 'Russia_150I', 'Russia_Km4I', 'Russia_VKPMI', 'UK_L232_type']\n"
     ]
    }
   ],
   "source": [
    "written.sort()\n",
    "keep_short.sort()\n",
    "#print(written)\n",
    "#print(keep_short)\n",
    "dropped.sort()\n",
    "print(dropped)\n",
    "#path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Mongolia-MGYG000240385</th>\n",
       "      <th>AM66-51ACA_human-feces_China_GCA_027667065</th>\n",
       "      <th>1001713B-A6_human-feces_NewYork_GCA_015558745</th>\n",
       "      <th>Mongolia-MGYG000100616</th>\n",
       "      <th>Mongolia-MGYG000012063</th>\n",
       "      <th>Bifidobacterium-adolescentis_AK_SK_36_Boto</th>\n",
       "      <th>UnitedStates-MGYG000241712</th>\n",
       "      <th>REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743</th>\n",
       "      <th>1001283B1-G1_human-feces_NewYork_GCA_015557595</th>\n",
       "      <th>...</th>\n",
       "      <th>1001275B-D1_human-feces_NewYork_GCA_015558565</th>\n",
       "      <th>CLA-AP-H11_human-feces_Germany_GCA_040092935</th>\n",
       "      <th>Mongolia-MGYG000175950</th>\n",
       "      <th>ElSalvador-MGYG000132829</th>\n",
       "      <th>Bifidobacterium-adolescentis_RS12_B_Warli</th>\n",
       "      <th>CLA-SR-H001_human-feces_Germany_GCA_040097245</th>\n",
       "      <th>AM64-16A_China_GCA_027667565</th>\n",
       "      <th>UnitedStates-MGYG000225709</th>\n",
       "      <th>REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ</th>\n",
       "      <th>Bifidobacterium-adolescentis_AK_SK_25_Purigpa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1200</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1371</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1695</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1802</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1816</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113300</th>\n",
       "      <td>40600</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113301</th>\n",
       "      <td>40686</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113302</th>\n",
       "      <td>41655</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113303</th>\n",
       "      <td>42041</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113304</th>\n",
       "      <td>42066</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113305 rows  133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0 Mongolia-MGYG000240385  \\\n",
       "0             1200                      A   \n",
       "1             1371                      T   \n",
       "2             1695                      G   \n",
       "3             1802                      T   \n",
       "4             1816                      G   \n",
       "...            ...                    ...   \n",
       "113300       40600                      C   \n",
       "113301       40686                      C   \n",
       "113302       41655                      C   \n",
       "113303       42041                      T   \n",
       "113304       42066                      G   \n",
       "\n",
       "       AM66-51ACA_human-feces_China_GCA_027667065  \\\n",
       "0                                               A   \n",
       "1                                               T   \n",
       "2                                               G   \n",
       "3                                               T   \n",
       "4                                               G   \n",
       "...                                           ...   \n",
       "113300                                          C   \n",
       "113301                                          C   \n",
       "113302                                          C   \n",
       "113303                                          T   \n",
       "113304                                          G   \n",
       "\n",
       "       1001713B-A6_human-feces_NewYork_GCA_015558745 Mongolia-MGYG000100616  \\\n",
       "0                                                  A                      A   \n",
       "1                                                  T                      T   \n",
       "2                                                  G                      G   \n",
       "3                                                  T                      T   \n",
       "4                                                  G                      G   \n",
       "...                                              ...                    ...   \n",
       "113300                                             C                      C   \n",
       "113301                                             C                      C   \n",
       "113302                                             C                      C   \n",
       "113303                                             T                      T   \n",
       "113304                                             G                      G   \n",
       "\n",
       "       Mongolia-MGYG000012063 Bifidobacterium-adolescentis_AK_SK_36_Boto  \\\n",
       "0                           A                                          A   \n",
       "1                           N                                          T   \n",
       "2                           N                                          G   \n",
       "3                           N                                          T   \n",
       "4                           N                                          G   \n",
       "...                       ...                                        ...   \n",
       "113300                      N                                          C   \n",
       "113301                      N                                          C   \n",
       "113302                      C                                          C   \n",
       "113303                      T                                          T   \n",
       "113304                      G                                          G   \n",
       "\n",
       "       UnitedStates-MGYG000241712  \\\n",
       "0                               A   \n",
       "1                               T   \n",
       "2                               G   \n",
       "3                               T   \n",
       "4                               G   \n",
       "...                           ...   \n",
       "113300                          C   \n",
       "113301                          C   \n",
       "113302                          C   \n",
       "113303                          T   \n",
       "113304                          G   \n",
       "\n",
       "       REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Hadza_MoBio_hadza-E-H_I_3_1743  \\\n",
       "0                                                       G                                          \n",
       "1                                                       C                                          \n",
       "2                                                       G                                          \n",
       "3                                                       A                                          \n",
       "4                                                       A                                          \n",
       "...                                                   ...                                          \n",
       "113300                                                  C                                          \n",
       "113301                                                  C                                          \n",
       "113302                                                  C                                          \n",
       "113303                                                  T                                          \n",
       "113304                                                  G                                          \n",
       "\n",
       "       1001283B1-G1_human-feces_NewYork_GCA_015557595  ...  \\\n",
       "0                                                   A  ...   \n",
       "1                                                   C  ...   \n",
       "2                                                   G  ...   \n",
       "3                                                   T  ...   \n",
       "4                                                   G  ...   \n",
       "...                                               ...  ...   \n",
       "113300                                              C  ...   \n",
       "113301                                              C  ...   \n",
       "113302                                              C  ...   \n",
       "113303                                              T  ...   \n",
       "113304                                              G  ...   \n",
       "\n",
       "       1001275B-D1_human-feces_NewYork_GCA_015558565  \\\n",
       "0                                                  A   \n",
       "1                                                  T   \n",
       "2                                                  G   \n",
       "3                                                  T   \n",
       "4                                                  G   \n",
       "...                                              ...   \n",
       "113300                                             C   \n",
       "113301                                             C   \n",
       "113302                                             C   \n",
       "113303                                             T   \n",
       "113304                                             G   \n",
       "\n",
       "       CLA-AP-H11_human-feces_Germany_GCA_040092935 Mongolia-MGYG000175950  \\\n",
       "0                                                 A                      A   \n",
       "1                                                 T                      T   \n",
       "2                                                 G                      G   \n",
       "3                                                 T                      T   \n",
       "4                                                 G                      G   \n",
       "...                                             ...                    ...   \n",
       "113300                                            C                      C   \n",
       "113301                                            C                      C   \n",
       "113302                                            C                      C   \n",
       "113303                                            T                      T   \n",
       "113304                                            G                      G   \n",
       "\n",
       "       ElSalvador-MGYG000132829 Bifidobacterium-adolescentis_RS12_B_Warli  \\\n",
       "0                             G                                         N   \n",
       "1                             C                                         T   \n",
       "2                             G                                         G   \n",
       "3                             T                                         T   \n",
       "4                             G                                         G   \n",
       "...                         ...                                       ...   \n",
       "113300                        C                                         C   \n",
       "113301                        C                                         C   \n",
       "113302                        C                                         C   \n",
       "113303                        T                                         T   \n",
       "113304                        G                                         G   \n",
       "\n",
       "       CLA-SR-H001_human-feces_Germany_GCA_040097245  \\\n",
       "0                                                  A   \n",
       "1                                                  T   \n",
       "2                                                  G   \n",
       "3                                                  T   \n",
       "4                                                  G   \n",
       "...                                              ...   \n",
       "113300                                             C   \n",
       "113301                                             C   \n",
       "113302                                             C   \n",
       "113303                                             T   \n",
       "113304                                             G   \n",
       "\n",
       "       AM64-16A_China_GCA_027667565 UnitedStates-MGYG000225709  \\\n",
       "0                                 A                          A   \n",
       "1                                 C                          T   \n",
       "2                                 G                          G   \n",
       "3                                 A                          T   \n",
       "4                                 A                          G   \n",
       "...                             ...                        ...   \n",
       "113300                            C                          C   \n",
       "113301                            C                          C   \n",
       "113302                            C                          C   \n",
       "113303                            T                          T   \n",
       "113304                            G                          G   \n",
       "\n",
       "       REFINED_METABAT215_TOP10_CONTIGS_1500_ASSEMBLY_K77_MERGED__Nepal_MoBio_Fiber-Hadza-Nepal_D_1_THA1067YZ  \\\n",
       "0                                                       A                                                       \n",
       "1                                                       T                                                       \n",
       "2                                                       G                                                       \n",
       "3                                                       T                                                       \n",
       "4                                                       G                                                       \n",
       "...                                                   ...                                                       \n",
       "113300                                                  C                                                       \n",
       "113301                                                  C                                                       \n",
       "113302                                                  C                                                       \n",
       "113303                                                  T                                                       \n",
       "113304                                                  G                                                       \n",
       "\n",
       "       Bifidobacterium-adolescentis_AK_SK_25_Purigpa  \n",
       "0                                                  A  \n",
       "1                                                  T  \n",
       "2                                                  G  \n",
       "3                                                  T  \n",
       "4                                                  G  \n",
       "...                                              ...  \n",
       "113300                                             C  \n",
       "113301                                             C  \n",
       "113302                                             C  \n",
       "113303                                             T  \n",
       "113304                                             G  \n",
       "\n",
       "[113305 rows x 133 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###########################################################################\n",
    "#######      032725: REWRITE PCA SNP TABLE AFTER DECISIONS TO REMOVE SAMPLES   - no \n",
    "###########################################################################\n",
    "\n",
    "# SNV_table_FILT_032526.csv, which is used to write PCA table, has columns (samples) I want to remove\n",
    "\n",
    "# then technically, the SNVs should be filtered again for usefuness of variation before the PCA table is written and MDS is done\n",
    "# this is less important for raxml because un-informative positions don't matter. \n",
    "# I think They actually dont really matter for  MDS either...but I just need accurate number of positions ...\n",
    "\n",
    "\n",
    "# The easiest way to remove samples is just delete the columns of this .csv in Excel on laptop \n",
    "\n",
    "# ## Read in info as pandas df where columns are easily removable \n",
    "# import pandas as pd\n",
    "# snvs = pd.read_csv(path+\"SNV_table_FILT_032526.csv\") # this is type anyway lol, should be 2025\n",
    "# #snvs\n",
    "\n",
    "# df_col_names = snvs.columns.tolist()\n",
    "# df_col_names.sort()\n",
    "\n",
    "# snvs = snvs.drop(columns=remove)\n",
    "# snvs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    }
   ],
   "source": [
    "remove = ['1-11_infant-feces_China_GCA_003030905','150_human-feces_Russia_GCA_001010915','ATCC-15703_dwnld031625_sbmtrJapan_GCA_000010425',\n",
    "\"Fiji-MGYG000104252\",\"Fiji-MGYG000124014\",\"Fiji-MGYG000199136\",\"Fiji-MGYG000248641\",\"Fiji-MGYG000255607\",\n",
    "\"Fiji-MGYG000285608\",\"Fiji-MGYG000031842\",\"Fiji-MGYG000076435\",\"Fiji-MGYG000098158\",\"Fiji-MGYG000099035\", \n",
    "\"JCM-15918_sbmtMongolia_GCA_000771705\",\"JCM15918_type-strain_NA_GCA_000741415\",'Km4_human-feces_Russia_GCA_001756865',\n",
    "\"L2-32_human-feces_NA_GCA_000154085\",\"LMG-10502_type_GCA_028743295\",\"LMG-10733_type_GCA_002108135\",\n",
    "\"Madagascar-MGYG000120594\",\"Madagascar-MGYG000012482\",\"Madagascar-MGYG000025145\",\"Madagascar-MGYG000091572\",\n",
    "\"NCTC11814_type_GCA_900445615\",\"VKPM-Ac-1245_human-feces_Russia_GCA_016069975\"]\n",
    "\n",
    "yes = 0\n",
    "for i in remove:\n",
    "    if i in df_col_names:\n",
    "        yes += 1\n",
    "    else:\n",
    "        print(i)\n",
    "print(yes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "#######       WRITE TO FILE FOR PCA -- correct version as of 032625 \n",
    "###########################################################################\n",
    "\n",
    "# Header (position + samples)\n",
    "# cells contain 1 for the rare allele, 0 for the common allele, or nothing if N\n",
    "\n",
    "#outfile = open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/Bif-adol_161_118975_PCA_102224.txt\",\"w\")\n",
    "\n",
    "# in 111824 version, remove replicate genomes by editing SNV_table_FILT_102224.csv in R ...\n",
    "# note this (R) fucks with the name,s but try anyway\n",
    "outfile = open(path+\"Bif-adol_157_113305_PCA_032625.txt\",\"w\")\n",
    "\n",
    "\n",
    "\n",
    "#df_transposed.to_csv('/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/SNV_table_102124.csv', index=True)\n",
    "with open(path+\"SNV_table_FILT_032526.csv\",\"r\") as csvfile:\n",
    "#with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/SNV_table_FILT_102224.csv\",\"r\") as csvfile:\n",
    "    reader = csv.reader(csvfile, delimiter=',') \n",
    "    header=next(reader)\n",
    "    header[0] = 'position'\n",
    "\n",
    "    for i in range(len(header)-1):\n",
    "        outfile.write(header[i]+',')\n",
    "    outfile.write(header[-1]+'\\n')\n",
    "        \n",
    "\n",
    "\n",
    "    for row in reader:\n",
    "        \n",
    "        outfile.write(row[0]+',')\n",
    "        data = row[1:len(row)]\n",
    "        \n",
    "        # replace N with blank\n",
    "        data = ['' if x=='N' else x for x in data]     \n",
    "        \n",
    "        # identify the common and rare allele and replace with 0/1\n",
    "        A_count = len([i for j,i in enumerate(data) if i =='A']) \n",
    "        T_count = len([i for j,i in enumerate(data) if i =='T'])\n",
    "        C_count = len([i for j,i in enumerate(data) if i =='C']) \n",
    "        G_count = len([i for j,i in enumerate(data) if i =='G'])\n",
    "\n",
    "        \n",
    "        if max(A_count,T_count,C_count,G_count) == A_count: # A is the ref allele\n",
    "            data = ['0' if x=='A' else x for x in data]\n",
    "            data = ['1' if x in ['T','C','G'] else x for x in data]\n",
    "            \n",
    "        elif max(A_count,T_count,C_count,G_count) == T_count: # T is the ref allele\n",
    "            data = ['0' if x=='T' else x for x in data]            \n",
    "            data = ['1' if x in ['A','C','G'] else x for x in data]\n",
    "            \n",
    "        elif max(A_count,T_count,C_count,G_count) == C_count: # C is the ref allele\n",
    "            data = ['0' if x=='C' else x for x in data]       \n",
    "            data = [\"1\" if x in ['T','A','G'] else x for x in data]\n",
    "        \n",
    "        elif max(A_count,T_count,C_count,G_count) == G_count: # G is the ref allele\n",
    "            data = ['0' if x=='G' else x for x in data]     \n",
    "            data = ['1' if x in ['T','C','A'] else x for x in data]\n",
    "        \n",
    "        \n",
    "        for i in range(len(data)-1):\n",
    "            outfile.write(data[i]+',')\n",
    "        outfile.write(data[-1]+'\\n')\n",
    "        \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ###########################################################################\n",
    "# #######       WRITE TO FILE FOR PCA -- correct version as of 032625 \n",
    "# ###########################################################################\n",
    "\n",
    "# # Header (position + samples)\n",
    "# # cells contain 1 for the rare allele, 0 for the common allele, or nothing if N\n",
    "\n",
    "# #outfile = open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/Bif-adol_161_118975_PCA_102224.txt\",\"w\")\n",
    "\n",
    "# # in 111824 version, remove replicate genomes by editing SNV_table_FILT_102224.csv in R ...\n",
    "# # note this (R) fucks with the name,s but try anyway\n",
    "# outfile = open(path+\"Bif-adol_157_113305_PCA_032625.txt\",\"w\")\n",
    "\n",
    "\n",
    "\n",
    "# #df_transposed.to_csv('/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/SNV_table_102124.csv', index=True)\n",
    "# with open(path+\"SNV_table_FILT_032526.csv\",\"r\") as csvfile:\n",
    "# #with open(\"/home/ebel/user_data/IndianMicrobiome/Bifidobacterium/adolescentis/mummer_files/ref_ATCC/SNV_table_FILT_102224.csv\",\"r\") as csvfile:\n",
    "#     reader = csv.reader(csvfile, delimiter=',') \n",
    "#     header=next(reader)\n",
    "#     header[0] = 'position'\n",
    "\n",
    "#     for i in range(len(header)-1):\n",
    "#         outfile.write(header[i]+',')\n",
    "#     outfile.write(header[-1]+'\\n')\n",
    "        \n",
    "\n",
    "\n",
    "#     for row in reader:\n",
    "        \n",
    "#         outfile.write(row[0]+',')\n",
    "#         data = row[1:len(row)]\n",
    "        \n",
    "#         # replace N with blank\n",
    "#         data = ['' if x=='N' else x for x in data]     \n",
    "        \n",
    "#         # identify the common and rare allele and replace with 0/1\n",
    "#         A_count = len([i for j,i in enumerate(data) if i =='A']) \n",
    "#         T_count = len([i for j,i in enumerate(data) if i =='T'])\n",
    "#         C_count = len([i for j,i in enumerate(data) if i =='C']) \n",
    "#         G_count = len([i for j,i in enumerate(data) if i =='G'])\n",
    "\n",
    "        \n",
    "#         if max(A_count,T_count,C_count,G_count) == A_count: # A is the ref allele\n",
    "#             data = ['0' if x=='A' else x for x in data]\n",
    "#             data = ['1' if x in ['T','C','G'] else x for x in data]\n",
    "            \n",
    "#         elif max(A_count,T_count,C_count,G_count) == T_count: # T is the ref allele\n",
    "#             data = ['0' if x=='T' else x for x in data]            \n",
    "#             data = ['1' if x in ['A','C','G'] else x for x in data]\n",
    "            \n",
    "#         elif max(A_count,T_count,C_count,G_count) == C_count: # C is the ref allele\n",
    "#             data = ['0' if x=='C' else x for x in data]       \n",
    "#             data = [\"1\" if x in ['T','A','G'] else x for x in data]\n",
    "        \n",
    "#         elif max(A_count,T_count,C_count,G_count) == G_count: # G is the ref allele\n",
    "#             data = ['0' if x=='G' else x for x in data]     \n",
    "#             data = ['1' if x in ['T','C','A'] else x for x in data]\n",
    "        \n",
    "        \n",
    "#         for i in range(len(data)-1):\n",
    "#             outfile.write(data[i]+',')\n",
    "#         outfile.write(data[-1]+'\\n')\n",
    "        \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "# outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
