This folder contains analysis scripts and input data tables from Ebel and Kulkarni et al., “‬Indian tribal gut microbiomes vary with dairy and grain consumption," in revision as of November 2025. 

Raw sequencing data are available at https://www.ncbi.nlm.nih.gov/bioproject/PRJNA1266865/. 



### Raw sequence analysis scripts 


- 16S_DADA2.R 
- metagenomics_Jupyter-notebooks.zip



### Input data tables 

- sample_metadata.csv
- Fig1_Diet_table_daily.csv
- Fig1_Diet_table_all.csv
- Fig2_16S_depth.csv
- Fig2_relab_RSV.csv (RSV = ribosomal sequencing variants (16S))
- Fig2_relab_genus16S.csv
- Fig2_relab_SRG.csv (SRG = species-representative genomes (metagenomics))


### Tables from other studies
- Tandon2018_GenusRelab.csv
- Wastyk2021_ASVrelab.rds
- Wastyk2021_ASVtaxonomy.rds
- Wastyk2021_metagenomics.csv


### Scripts producing figures 

- Fig1_diet.R 
- Fig2_alpha.R
- Fig2_PCA.R
- Fig2_outliers.R
- Fig2_Supp2_StackedBar_Individual_16S.R
- Fig2_Supp3_StackedBar_Individual_metagenomics.R
- Fig3_core.R
- Fig3_Supp1_VANISH.R
- Fig4_Maaslin.R
- Fig4_StackedBar_Tribe.R
- Fig5_totals.R
- Fig5_stackedbars.R
- Fig6_MDS.R
- Fig6D.R
- Fig7_CAZy.R
- Fig7B.R

### Intermediate analysis scripts and tables 
- mummer_allele_table_032525.ipynb (Parses Bifidobacterium adolescentis SNVs obtained with nucmer)
- dbcan.zip (Contains overview.txt files generated by running dbcan2 on bacterial genomes) 
- CAZymes_by_genome_103025.csv
- CAZymes_by_genome_PERMB_103025.csv
- CAZymes_by_sample_IMplusFFF_103025.csv

